{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Source import data_chanel_map\n",
    "from Source import forward_color\n",
    "from Source import reverse_color\n",
    "from Source import reverse_map\n",
    "from Source import Definition\n",
    "\n",
    "definition = Definition(\"./resource/tables.pdata\")\n",
    "\n",
    "class BinFile:\n",
    "    def __init__(self, file) -> None:\n",
    "        self.data = self.read_file(file)\n",
    "        self.process_data = self.parser()\n",
    "\n",
    "    def read_file(self, file):\n",
    "        with open(file, 'r') as f:\n",
    "            return [line.strip('\\n') for line in f.readlines()]\n",
    "\n",
    "    def parser(self):\n",
    "        lines = []\n",
    "        for line in self.data:\n",
    "            # print(line)\n",
    "            if 'Version' in line:\n",
    "                version = line.split('\\t')[1]\n",
    "            elif 'Chemistry' in line:\n",
    "                kit_name = line.split('\\t')[1]\n",
    "            elif 'Panel Name' in line:\n",
    "                panel_name = line.split('\\t')[1]\n",
    "            elif 'Marker Name' in line:\n",
    "                marker_name = line.split('\\t')[1]\n",
    "            elif 'BinSet' in line:\n",
    "                continue\n",
    "            else:\n",
    "                # print(panel_name)\n",
    "                line_base = line.split('\\t')\n",
    "                line_base.insert(0, marker_name)\n",
    "                line_base.insert(0, panel_name)\n",
    "                line_base.insert(0, kit_name)\n",
    "                line_base.insert(0, version)\n",
    "                lines.append(line_base)\n",
    "\n",
    "        \n",
    "        df = pd.DataFrame(lines, columns=['version', 'package', 'defined_panel', 'marker', 'base', 'min_bin', 'max_bin', 'color'])\n",
    "        df = df[pd.notnull(df.color)]\n",
    "        df['min_bin'] = round(df['min_bin'].astype('float'),2)\n",
    "        df['max_bin'] = round(df['max_bin'].astype('float'),2)\n",
    "        df['color'] = df['color'].apply(lambda x: x.lower())\n",
    "        \n",
    "        # conver yellow to black\n",
    "        df['color'] = df['color'].apply(lambda x: 'black' if x  == 'yellow' else x)\n",
    "        \n",
    "        # update direction\n",
    "        df['direction'] = df.apply(lambda row: self.update_direction(row['base'], row['color']), axis=1)\n",
    "\n",
    "        # update is_forward\n",
    "        df['is_forward'] = df['direction'].apply(lambda x: 1 if x == 'Forward' else 0)\n",
    "\n",
    "        # update label\n",
    "        df = df.merge(definition.marker_table[['marker', 'marker_label', 'gene']], how='left')\n",
    "\n",
    "        # update panel\n",
    "        df['panel'] = df['defined_panel'].apply(lambda x: f'S{x[-1]}')\n",
    "\n",
    "        # update basetype\n",
    "\n",
    "        def _base_type(marker, base):\n",
    "            marker_data = definition.marker_table[definition.marker_table['marker'] ==  marker]\n",
    "            wildtype = marker_data['wildtype'].to_list()[0]\n",
    "\n",
    "            return 'wildtype' if base == wildtype else 'mutant'\n",
    "        df['basetype'] = df.apply(lambda row: _base_type(row['marker'], row['base']), axis=1)\n",
    "                \n",
    "\n",
    "        # intiate min intensity\n",
    "        df['min_height'] = 1000\n",
    "\n",
    "        # process duplicate\n",
    "        # get max range file of bin setting if base duplicated\n",
    "        df['binrange'] = df.max_bin - df.min_bin\n",
    "\n",
    "        # get max bin range\n",
    "        df = df.sort_values('binrange', ascending=False).groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).first()\n",
    "\n",
    "        df.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True], inplace=True)\n",
    "\n",
    "        # reorder colums\n",
    "        df = df[['package', 'version', 'defined_panel', 'panel', 'gene', 'marker',  'marker_label', 'is_forward',  'direction', \n",
    "                'base', 'basetype', 'color',  'min_height', 'min_bin', 'max_bin']]\n",
    "\n",
    "\n",
    "        # check status marker\n",
    "\n",
    "        markers = df[['panel','marker', 'direction']].drop_duplicates()\n",
    "        count = markers.groupby('marker').size()\n",
    "\n",
    "        if any(count[count > 1]):\n",
    "            error_markers = \"; \".join(count[count > 1].index)\n",
    "            raise ValueError(f'Marker(s) info not consistent [{error_markers}]')\n",
    "        \n",
    "        else:\n",
    "            return df\n",
    "    \n",
    "    def update_direction(self, base, color):\n",
    "        return 'Forward' if forward_color.get(base) == color else 'Reverse'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "bin_file = './validation_panels/panels/[2021.08 └┌╖ß] PGx-NP_PGx-NP_bins_lam_updated.txt'\n",
    "bin = BinFile(bin_file)\n",
    "bin.process_data.to_excel('./validation_panels/PGx_NP_bin_seting_table_new_form.xlsx', index=False)\n",
    "data = bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>version</th>\n",
       "      <th>defined_panel</th>\n",
       "      <th>panel</th>\n",
       "      <th>gene</th>\n",
       "      <th>marker</th>\n",
       "      <th>marker_label</th>\n",
       "      <th>is_forward</th>\n",
       "      <th>direction</th>\n",
       "      <th>base</th>\n",
       "      <th>basetype</th>\n",
       "      <th>color</th>\n",
       "      <th>min_height</th>\n",
       "      <th>min_bin</th>\n",
       "      <th>max_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>CYP2D6 *49</td>\n",
       "      <td>0</td>\n",
       "      <td>Reverse</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>37.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.5</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    package   version defined_panel panel    gene      marker   marker_label  \\\n",
       "1    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "0    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "2    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "3    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "5    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_003     CYP2D6 *49   \n",
       "..      ...       ...           ...   ...     ...         ...            ...   \n",
       "92   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "93   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "106  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "105  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "107  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "\n",
       "     is_forward direction base  basetype  color  min_height  min_bin  max_bin  \n",
       "1             1   Forward    G  wildtype   blue        1000     25.0     35.0  \n",
       "0             1   Forward    A    mutant  green        1000     27.0     36.0  \n",
       "2             1   Forward    C  wildtype  black        1000     28.0     38.0  \n",
       "3             1   Forward    T    mutant    red        1000     31.0     37.0  \n",
       "5             0   Reverse    T  wildtype  green        1000     37.0     44.0  \n",
       "..          ...       ...  ...       ...    ...         ...      ...      ...  \n",
       "92            1   Forward    A    mutant  green        1000     68.0     75.0  \n",
       "93            1   Forward    C  wildtype  black        1000     68.5     75.0  \n",
       "106           1   Forward    C  wildtype  black        1000     76.0     84.0  \n",
       "105           1   Forward    A    mutant  green        1000     76.0     83.0  \n",
       "107           1   Forward    T    mutant    red        1000     76.0     84.0  \n",
       "\n",
       "[112 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from Source import generate_markers\n",
    "import pandas as pd\n",
    "data = pd.read_excel(\"validation_panels/PGx_NP_bin_seting_table_new_form.xlsx\")\n",
    "markers = generate_markers(data, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].defined_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].color"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('snapshot_genotyping')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b4d3fe072432f0816511cf4002824cbf67c2cdbd34bdb0afb8cba97d824c69ea"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
