{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Source import data_chanel_map\n",
    "from Source import forward_color\n",
    "from Source import reverse_color\n",
    "from Source import reverse_map\n",
    "from Source import Definition\n",
    "\n",
    "definition = Definition(\"./resource/tables.pdata\")\n",
    "\n",
    "class BinFile:\n",
    "    def __init__(self, file) -> None:\n",
    "        self.data = self.read_file(file)\n",
    "        self.process_data = self.parser()\n",
    "\n",
    "    def read_file(self, file):\n",
    "        with open(file, 'r') as f:\n",
    "            return [line.strip('\\n') for line in f.readlines()]\n",
    "\n",
    "    def parser(self):\n",
    "        lines = []\n",
    "        for line in self.data:\n",
    "            # print(line)\n",
    "            if 'Version' in line:\n",
    "                version = line.split('\\t')[1]\n",
    "            elif 'Chemistry' in line:\n",
    "                kit_name = line.split('\\t')[1]\n",
    "            elif 'Panel Name' in line:\n",
    "                panel_name = line.split('\\t')[1]\n",
    "            elif 'Marker Name' in line:\n",
    "                marker_name = line.split('\\t')[1]\n",
    "            elif 'BinSet' in line:\n",
    "                continue\n",
    "            else:\n",
    "                # print(panel_name)\n",
    "                line_base = line.split('\\t')\n",
    "                line_base.insert(0, marker_name)\n",
    "                line_base.insert(0, panel_name)\n",
    "                line_base.insert(0, kit_name)\n",
    "                line_base.insert(0, version)\n",
    "                lines.append(line_base)\n",
    "\n",
    "        \n",
    "        df = pd.DataFrame(lines, columns=['version', 'package', 'defined_panel', 'marker', 'base', 'min_bin', 'max_bin', 'color'])\n",
    "        df = df[pd.notnull(df.color)]\n",
    "        df['min_bin'] = round(df['min_bin'].astype('float'),2)\n",
    "        df['max_bin'] = round(df['max_bin'].astype('float'),2)\n",
    "        df['color'] = df['color'].apply(lambda x: x.lower())\n",
    "        \n",
    "        # conver yellow to black\n",
    "        df['color'] = df['color'].apply(lambda x: 'black' if x  == 'yellow' else x)\n",
    "        \n",
    "        # update direction\n",
    "        df['direction'] = df.apply(lambda row: self.update_direction(row['base'], row['color']), axis=1)\n",
    "\n",
    "        # update is_forward\n",
    "        df['is_forward'] = df['direction'].apply(lambda x: 1 if x == 'Forward' else 0)\n",
    "\n",
    "        # update label\n",
    "        df = df.merge(definition.marker_table[['marker', 'marker_label', 'gene']], how='left')\n",
    "\n",
    "        # update panel\n",
    "        df['panel'] = df['defined_panel'].apply(lambda x: f'S{x[-1]}')\n",
    "\n",
    "        # update basetype\n",
    "\n",
    "        def _base_type(marker, base):\n",
    "            # print(marker)\n",
    "            marker_data = definition.marker_table[definition.marker_table['marker'] ==  marker]\n",
    "            print(marker_data)\n",
    "            wildtype = marker_data['wildtype'].to_list()[0]\n",
    "\n",
    "            return 'wildtype' if base == wildtype else 'mutant'\n",
    "\n",
    "\n",
    "        df['basetype'] = df.apply(lambda row: _base_type(row['marker'], row['base']), axis=1)\n",
    "                \n",
    "\n",
    "        # intiate min intensity\n",
    "        df['min_height'] = 1000\n",
    "\n",
    "        # process duplicate\n",
    "        # get max range file of bin setting if base duplicated\n",
    "        # M1 20  25\n",
    "        # M1 21  26\n",
    "        # --> M1 20 26 \n",
    "\n",
    "        tmp = df.groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).agg(bin_min=('min_bin', 'min'), bin_max=('max_bin', 'max'))\n",
    "\n",
    "        df = df.merge(tmp)\n",
    "\n",
    "        df.drop(columns=['min_bin', 'max_bin'], inplace=True)\n",
    "        df.rename(columns={'bin_min':'min_bin', 'bin_max':'max_bin'}, inplace=True)\n",
    "\n",
    "        # commennt the above code if we want to get the max bin range instead\n",
    "\n",
    "        df['binrange'] = df.min_bin - df.max_bin\n",
    "\n",
    "        # get max bin range\n",
    "        df = df.sort_values('binrange', ascending=False).groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).first()\n",
    "\n",
    "        df.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True], inplace=True)\n",
    "       \n",
    "        # reorder colums\n",
    "        df = df[['package', 'version', 'defined_panel', 'panel', 'gene', 'marker',  'marker_label', 'is_forward',  'direction', \n",
    "                'base', 'basetype', 'color',  'min_height', 'min_bin', 'max_bin']]\n",
    "\n",
    "\n",
    "        # check status marker\n",
    "\n",
    "        markers = df[['panel','marker', 'direction']].drop_duplicates()\n",
    "        count = markers.groupby('marker').size()\n",
    "\n",
    "        if any(count[count > 1]):\n",
    "            error_markers = \"; \".join(count[count > 1].index)\n",
    "            raise ValueError(f'Marker(s) info not consistent [{error_markers}]')\n",
    "        \n",
    "        else:\n",
    "            return df\n",
    "    \n",
    "    def update_direction(self, base, color):\n",
    "        return 'Forward' if forward_color.get(base) == color else 'Reverse'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "20  CYP2D6_001         *14     G169R   1758G>A  rs5030865        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "20       14         21  CYP2D6   CYP2D6 *14  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "20  CYP2D6_001         *14     G169R   1758G>A  rs5030865        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "20       14         21  CYP2D6   CYP2D6 *14  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "21  CYP2D6_002        *10B      P34S    100C>T  rs1065852        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "21       14         22  CYP2D6  CYP2D6 *10B  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "21  CYP2D6_002        *10B      P34S    100C>T  rs1065852        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "21       14         22  CYP2D6  CYP2D6 *10B  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "22  CYP2D6_003         *49     F120I   1611T>A  rs1135822        T      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "22       14         23  CYP2D6   CYP2D6 *49  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "22  CYP2D6_003         *49     F120I   1611T>A  rs1135822        T      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "22       14         23  CYP2D6   CYP2D6 *49  \n",
      "        marker star_allele       aa_change      nt_change        rsid  \\\n",
      "23  CYP2D6_004         *21  267_frameshift  2573_2574insC  rs72549352   \n",
      "\n",
      "   wildtype mutant  gene_id  marker_id    gene marker_label  \n",
      "23        G      C       14         24  CYP2D6   CYP2D6 *21  \n",
      "        marker star_allele       aa_change      nt_change        rsid  \\\n",
      "23  CYP2D6_004         *21  267_frameshift  2573_2574insC  rs72549352   \n",
      "\n",
      "   wildtype mutant  gene_id  marker_id    gene marker_label  \n",
      "23        G      C       14         24  CYP2D6   CYP2D6 *21  \n",
      "        marker star_allele        aa_change nt_change        rsid wildtype  \\\n",
      "24  CYP2D6_005         *41  Splicing_defect   2988G>A  rs28371725        G   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "24      A       14         25  CYP2D6   CYP2D6 *41  \n",
      "        marker star_allele        aa_change nt_change        rsid wildtype  \\\n",
      "24  CYP2D6_005         *41  Splicing_defect   2988G>A  rs28371725        G   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "24      A       14         25  CYP2D6   CYP2D6 *41  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "25  CYP2D6_006         *52     E418K   3877G>A  rs28371733        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "25       14         26  CYP2D6   CYP2D6 *52  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "25  CYP2D6_006         *52     E418K   3877G>A  rs28371733        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "25       14         26  CYP2D6   CYP2D6 *52  \n",
      "        marker star_allele      aa_change               nt_change rsid  \\\n",
      "26  CYP2D6_007         *18  468_470dupVPT  4125_4133_dupGTGCCCACT    -   \n",
      "\n",
      "   wildtype mutant  gene_id  marker_id    gene marker_label  \n",
      "26        G      T       14         27  CYP2D6   CYP2D6 *18  \n",
      "        marker star_allele      aa_change               nt_change rsid  \\\n",
      "26  CYP2D6_007         *18  468_470dupVPT  4125_4133_dupGTGCCCACT    -   \n",
      "\n",
      "   wildtype mutant  gene_id  marker_id    gene marker_label  \n",
      "26        G      T       14         27  CYP2D6   CYP2D6 *18  \n",
      "        marker star_allele aa_change nt_change     rsid wildtype mutant  \\\n",
      "27  CYP2D6_008          *2     R296C   2850C>T  rs16947        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "27       14         28  CYP2D6    CYP2D6 *2  \n",
      "        marker star_allele aa_change nt_change     rsid wildtype mutant  \\\n",
      "27  CYP2D6_008          *2     R296C   2850C>T  rs16947        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "27       14         28  CYP2D6    CYP2D6 *2  \n",
      "        marker star_allele aa_change  nt_change rsid wildtype mutant  gene_id  \\\n",
      "28  CYP2D6_009         *60     S183X  1887insTA    -        G      A       14   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "28         29  CYP2D6   CYP2D6 *60  \n",
      "        marker star_allele aa_change  nt_change rsid wildtype mutant  gene_id  \\\n",
      "28  CYP2D6_009         *60     S183X  1887insTA    -        G      A       14   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "28         29  CYP2D6   CYP2D6 *60  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "29  CYP2D6_010          *5         -  Deletion    -        A      G       14   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "29         30  CYP2D6    CYP2D6 *5  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "29  CYP2D6_010          *5         -  Deletion    -        A      G       14   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "29         30  CYP2D6    CYP2D6 *5  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "54  HLA-A_004           -         -  593T>C/A    -        T    C/A       25   \n",
      "\n",
      "    marker_id   gene    marker_label  \n",
      "54         55  HLA-A  HLA-A 593T>C/A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "54  HLA-A_004           -         -  593T>C/A    -        T    C/A       25   \n",
      "\n",
      "    marker_id   gene    marker_label  \n",
      "54         55  HLA-A  HLA-A 593T>C/A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "54  HLA-A_004           -         -  593T>C/A    -        T    C/A       25   \n",
      "\n",
      "    marker_id   gene    marker_label  \n",
      "54         55  HLA-A  HLA-A 593T>C/A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "55  HLA-B_001           -         -    384C>G    -        C      G       26   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "55         56  HLA-B  HLA-B 384C>G  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "55  HLA-B_001           -         -    384C>G    -        C      G       26   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "55         56  HLA-B  HLA-B 384C>G  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "56  HLA-B_002           -         -    667G>A    -        G      A       26   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "56         57  HLA-B  HLA-B 667G>A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "56  HLA-B_002           -         -    667G>A    -        G      A       26   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "56         57  HLA-B  HLA-B 667G>A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "57  HLA-B_003           -         -   2156A>T    -        A      T       26   \n",
      "\n",
      "    marker_id   gene   marker_label  \n",
      "57         58  HLA-B  HLA-B 2156A>T  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "57  HLA-B_003           -         -   2156A>T    -        A      T       26   \n",
      "\n",
      "    marker_id   gene   marker_label  \n",
      "57         58  HLA-B  HLA-B 2156A>T  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "51  HLA-A_001           -         -   368G>A     -        G      A       25   \n",
      "\n",
      "    marker_id   gene   marker_label  \n",
      "51         52  HLA-A  HLA-A 368G>A   \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "51  HLA-A_001           -         -   368G>A     -        G      A       25   \n",
      "\n",
      "    marker_id   gene   marker_label  \n",
      "51         52  HLA-A  HLA-A 368G>A   \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "52  HLA-A_002           -         -    954T>C    -        T      C       25   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "52         53  HLA-A  HLA-A 954T>C  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "52  HLA-A_002           -         -    954T>C    -        T      C       25   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "52         53  HLA-A  HLA-A 954T>C  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "53  HLA-A_003           -         -    536G>A    -        G      A       25   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "53         54  HLA-A  HLA-A 536G>A  \n",
      "       marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "53  HLA-A_003           -         -    536G>A    -        G      A       25   \n",
      "\n",
      "    marker_id   gene  marker_label  \n",
      "53         54  HLA-A  HLA-A 536G>A  \n",
      "        marker star_allele        aa_change     nt_change      rsid wildtype  \\\n",
      "39  CYP3A5_001          *3  splicing defect  c.219-237A>G  rs776746        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "39      G       16         40  CYP3A5    CYP3A5 *3  \n",
      "        marker star_allele        aa_change     nt_change      rsid wildtype  \\\n",
      "39  CYP3A5_001          *3  splicing defect  c.219-237A>G  rs776746        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "39      G       16         40  CYP3A5    CYP3A5 *3  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "         marker star_allele aa_change  nt_change        rsid wildtype mutant  \\\n",
      "13  CYP2C19_001         *17     5'UTR  c.-806C>T  rs12248560        C      T   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "13       12         14  CYP2C19  CYP2C19 *17  \n",
      "        marker star_allele aa_change   nt_change        rsid wildtype mutant  \\\n",
      "38  CYP3A4_002         *22         -  g.15389C>T  rs35599367        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "38       15         39  CYP3A4   CYP3A4 *22  \n",
      "        marker star_allele aa_change   nt_change        rsid wildtype mutant  \\\n",
      "38  CYP3A4_002         *22         -  g.15389C>T  rs35599367        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "38       15         39  CYP3A4   CYP3A4 *22  \n",
      "        marker star_allele aa_change   nt_change        rsid wildtype mutant  \\\n",
      "38  CYP3A4_002         *22         -  g.15389C>T  rs35599367        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "38       15         39  CYP3A4   CYP3A4 *22  \n",
      "        marker star_allele aa_change   nt_change        rsid wildtype mutant  \\\n",
      "38  CYP3A4_002         *22         -  g.15389C>T  rs35599367        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "38       15         39  CYP3A4   CYP3A4 *22  \n",
      "        marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "16  CYP2C9_001          *3     I359L  c.1075A>C  rs1057910        A      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "16       13         17  CYP2C9    CYP2C9 *3  \n",
      "        marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "16  CYP2C9_001          *3     I359L  c.1075A>C  rs1057910        A      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "16       13         17  CYP2C9    CYP2C9 *3  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "17  CYP2C9_002         *13      L90P  c.269T>C  rs72558187        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "17       13         18  CYP2C9   CYP2C9 *13  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "17  CYP2C9_002         *13      L90P  c.269T>C  rs72558187        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "17       13         18  CYP2C9   CYP2C9 *13  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "18  CYP2C9_003         *14     R125H  c.374G>A  rs72558189        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "18       13         19  CYP2C9   CYP2C9 *14  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "18  CYP2C9_003         *14     R125H  c.374G>A  rs72558189        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "18       13         19  CYP2C9   CYP2C9 *14  \n",
      "         marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "15  CYP2C19_003          *3     W212X  c.636G>A  rs4986893        G      A   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "15       12         16  CYP2C19   CYP2C19 *3  \n",
      "         marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "15  CYP2C19_003          *3     W212X  c.636G>A  rs4986893        G      A   \n",
      "\n",
      "    gene_id  marker_id     gene marker_label  \n",
      "15       12         16  CYP2C19   CYP2C19 *3  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "11  CYP2B6_001          *4     K262R  18053A>G  rs2279343        A      G   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "11       11         12  CYP2B6    CYP2B6 *4  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "11  CYP2B6_001          *4     K262R  18053A>G  rs2279343        A      G   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "11       11         12  CYP2B6    CYP2B6 *4  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "12  CYP2B6_002          *9     Q172H  15631G>T  rs3745274        G      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "12       11         13  CYP2B6    CYP2B6 *9  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "12  CYP2B6_002          *9     Q172H  15631G>T  rs3745274        G      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "12       11         13  CYP2B6    CYP2B6 *9  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "19  CYP2C9_004          *2     R144C  c.430C>T  rs1799853        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "19       13         20  CYP2C9    CYP2C9 *2  \n",
      "        marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "19  CYP2C9_004          *2     R144C  c.430C>T  rs1799853        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "19       13         20  CYP2C9    CYP2C9 *2  \n",
      "         marker star_allele        aa_change nt_change       rsid wildtype  \\\n",
      "14  CYP2C19_002          *2  splicing defect  c.681G>A  rs4244285        G   \n",
      "\n",
      "   mutant  gene_id  marker_id     gene marker_label  \n",
      "14      A       12         15  CYP2C19   CYP2C19 *2  \n",
      "         marker star_allele        aa_change nt_change       rsid wildtype  \\\n",
      "14  CYP2C19_002          *2  splicing defect  c.681G>A  rs4244285        G   \n",
      "\n",
      "   mutant  gene_id  marker_id     gene marker_label  \n",
      "14      A       12         15  CYP2C19   CYP2C19 *2  \n",
      "       marker star_allele aa_change      nt_change    rsid wildtype mutant  \\\n",
      "58  HTR1A_001           -         -  g.63962738C>G  rs6295        C      G   \n",
      "\n",
      "    gene_id  marker_id   gene  marker_label  \n",
      "58       27         59  HTR1A  HTR1A rs6295  \n",
      "       marker star_allele aa_change      nt_change    rsid wildtype mutant  \\\n",
      "58  HTR1A_001           -         -  g.63962738C>G  rs6295        C      G   \n",
      "\n",
      "    gene_id  marker_id   gene  marker_label  \n",
      "58       27         59  HTR1A  HTR1A rs6295  \n",
      "       marker star_allele aa_change    nt_change       rsid wildtype mutant  \\\n",
      "60  HTR2C_001           -         -  g.324497C>G  rs1414334        C      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "60       29         61  HTR2C  HTR2C rs1414334  \n",
      "       marker star_allele aa_change    nt_change       rsid wildtype mutant  \\\n",
      "60  HTR2C_001           -         -  g.324497C>G  rs1414334        C      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "60       29         61  HTR2C  HTR2C rs1414334  \n",
      "     marker star_allele aa_change nt_change    rsid wildtype mutant  gene_id  \\\n",
      "8  COMT_001           -     V158M  c.472G>A  rs4680        G      A        8   \n",
      "\n",
      "   marker_id  gene marker_label  \n",
      "8          9  COMT  COMT rs4680  \n",
      "     marker star_allele aa_change nt_change    rsid wildtype mutant  gene_id  \\\n",
      "8  COMT_001           -     V158M  c.472G>A  rs4680        G      A        8   \n",
      "\n",
      "   marker_id  gene marker_label  \n",
      "8          9  COMT  COMT rs4680  \n",
      "      marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "1  ABCB1_002           -    I1145I  c.3435A>G  rs1045642        A      G   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "1        1          2  ABCB1  ABCB1 rs1045642  \n",
      "      marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "1  ABCB1_002           -    I1145I  c.3435A>G  rs1045642        A      G   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "1        1          2  ABCB1  ABCB1 rs1045642  \n",
      "      marker star_allele aa_change        nt_change       rsid wildtype  \\\n",
      "41  DRD2_001           -         -  c.-486_-485insC  rs1799732        -   \n",
      "\n",
      "   mutant  gene_id  marker_id  gene    marker_label  \n",
      "41      G       18         42  DRD2  DRD2 rs1799732  \n",
      "      marker star_allele aa_change        nt_change       rsid wildtype  \\\n",
      "41  DRD2_001           -         -  c.-486_-485insC  rs1799732        -   \n",
      "\n",
      "   mutant  gene_id  marker_id  gene    marker_label  \n",
      "41      G       18         42  DRD2  DRD2 rs1799732  \n",
      "      marker star_allele aa_change    nt_change       rsid wildtype mutant  \\\n",
      "0  ABCB1_001           -    A893ST  c.2677A>C/T  rs2032582        A    C/T   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "0        1          1  ABCB1  ABCB1 rs2032582  \n",
      "      marker star_allele aa_change    nt_change       rsid wildtype mutant  \\\n",
      "0  ABCB1_001           -    A893ST  c.2677A>C/T  rs2032582        A    C/T   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "0        1          1  ABCB1  ABCB1 rs2032582  \n",
      "      marker star_allele aa_change    nt_change       rsid wildtype mutant  \\\n",
      "0  ABCB1_001           -    A893ST  c.2677A>C/T  rs2032582        A    C/T   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "0        1          1  ABCB1  ABCB1 rs2032582  \n",
      "      marker star_allele aa_change nt_change    rsid wildtype mutant  gene_id  \\\n",
      "42  DRD3_001           -       G9S   c.25G>A  rs6280        G      A       19   \n",
      "\n",
      "    marker_id  gene marker_label  \n",
      "42         43  DRD3  DRD3 rs6280  \n",
      "      marker star_allele aa_change nt_change    rsid wildtype mutant  gene_id  \\\n",
      "42  DRD3_001           -       G9S   c.25G>A  rs6280        G      A       19   \n",
      "\n",
      "    marker_id  gene marker_label  \n",
      "42         43  DRD3  DRD3 rs6280  \n",
      "       marker star_allele aa_change       nt_change       rsid wildtype  \\\n",
      "70  SCN1A_001           -         -  g.166053034C>T  rs3812718        C   \n",
      "\n",
      "   mutant  gene_id  marker_id   gene     marker_label  \n",
      "70      T       37         71  SCN1A  SCN1A rs3812718  \n",
      "       marker star_allele aa_change       nt_change       rsid wildtype  \\\n",
      "70  SCN1A_001           -         -  g.166053034C>T  rs3812718        C   \n",
      "\n",
      "   mutant  gene_id  marker_id   gene     marker_label  \n",
      "70      T       37         71  SCN1A  SCN1A rs3812718  \n",
      "      marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "6  ANKK1_001           -     E713K  c.2137G>A  rs1800497        G      A   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "6        6          7  ANKK1  ANKK1 rs1800497  \n",
      "      marker star_allele aa_change  nt_change       rsid wildtype mutant  \\\n",
      "6  ANKK1_001           -     E713K  c.2137G>A  rs1800497        G      A   \n",
      "\n",
      "   gene_id  marker_id   gene     marker_label  \n",
      "6        6          7  ANKK1  ANKK1 rs1800497  \n",
      "       marker star_allele aa_change   nt_change       rsid wildtype mutant  \\\n",
      "45  EPHX1_001           -   H139R/L  c.416A>G/T  rs2234922        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "45       21         46  EPHX1  EPHX1 rs2234922  \n",
      "       marker star_allele aa_change   nt_change       rsid wildtype mutant  \\\n",
      "45  EPHX1_001           -   H139R/L  c.416A>G/T  rs2234922        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "45       21         46  EPHX1  EPHX1 rs2234922  \n",
      "       marker star_allele aa_change   nt_change       rsid wildtype mutant  \\\n",
      "45  EPHX1_001           -   H139R/L  c.416A>G/T  rs2234922        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "45       21         46  EPHX1  EPHX1 rs2234922  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "37  CYP3A4_001         *18     L293P    878T>C  rs28371759        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "37       15         38  CYP3A4   CYP3A4 *18  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "37  CYP3A4_001         *18     L293P    878T>C  rs28371759        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "37       15         38  CYP3A4   CYP3A4 *18  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "37  CYP3A4_001         *18     L293P    878T>C  rs28371759        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "37       15         38  CYP3A4   CYP3A4 *18  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "37  CYP3A4_001         *18     L293P    878T>C  rs28371759        T      C   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "37       15         38  CYP3A4   CYP3A4 *18  \n",
      "       marker star_allele aa_change      nt_change       rsid wildtype mutant  \\\n",
      "47  FKBP5_001           -         -  g.35702206A>G  rs4713916        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "47       22         48  FKBP5  FKBP5 rs4713916  \n",
      "       marker star_allele aa_change      nt_change       rsid wildtype mutant  \\\n",
      "47  FKBP5_001           -         -  g.35702206A>G  rs4713916        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "47       22         48  FKBP5  FKBP5 rs4713916  \n",
      "       marker star_allele aa_change      nt_change       rsid wildtype mutant  \\\n",
      "59  HTR2A_001           -         -  g.46837850A>G  rs7997012        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "59       28         60  HTR2A  HTR2A rs7997012  \n",
      "       marker star_allele aa_change      nt_change       rsid wildtype mutant  \\\n",
      "59  HTR2A_001           -         -  g.46837850A>G  rs7997012        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "59       28         60  HTR2A  HTR2A rs7997012  \n",
      "         marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "71  UGT2B15_001           -      D85Y    253G>T    -        G      T       38   \n",
      "\n",
      "    marker_id     gene    marker_label  \n",
      "71         72  UGT2B15  UGT2B15 253G>T  \n",
      "         marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "71  UGT2B15_001           -      D85Y    253G>T    -        G      T       38   \n",
      "\n",
      "    marker_id     gene    marker_label  \n",
      "71         72  UGT2B15  UGT2B15 253G>T  \n",
      "       marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "46  EPHX1_002           -     Y113H  c.337T>C  rs1051740        T      C   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "46       21         47  EPHX1  EPHX1 rs1051740  \n",
      "       marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "46  EPHX1_002           -     Y113H  c.337T>C  rs1051740        T      C   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "46       21         47  EPHX1  EPHX1 rs1051740  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "72  UGT1A4_001           -      L48V    142T>G    -        T      G       39   \n",
      "\n",
      "    marker_id    gene   marker_label  \n",
      "72         73  UGT1A4  UGT1A4 142T>G  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "72  UGT1A4_001           -      L48V    142T>G    -        T      G       39   \n",
      "\n",
      "    marker_id    gene   marker_label  \n",
      "72         73  UGT1A4  UGT1A4 142T>G  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "10  CYP1A2_001         *1F         -   -163C>A    -        C      A       10   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "10         11  CYP1A2   CYP1A2 *1F  \n",
      "        marker star_allele aa_change nt_change rsid wildtype mutant  gene_id  \\\n",
      "10  CYP1A2_001         *1F         -   -163C>A    -        C      A       10   \n",
      "\n",
      "    marker_id    gene marker_label  \n",
      "10         11  CYP1A2   CYP1A2 *1F  \n",
      "      marker star_allele aa_change        nt_change      rsid wildtype mutant  \\\n",
      "62  MC4R_001           -         -  g.57882787C>A/T  rs489693        C      A   \n",
      "\n",
      "    gene_id  marker_id  gene   marker_label  \n",
      "62       31         63  MC4R  MC4R rs489693  \n",
      "      marker star_allele aa_change        nt_change      rsid wildtype mutant  \\\n",
      "62  MC4R_001           -         -  g.57882787C>A/T  rs489693        C      A   \n",
      "\n",
      "    gene_id  marker_id  gene   marker_label  \n",
      "62       31         63  MC4R  MC4R rs489693  \n",
      "      marker star_allele aa_change        nt_change      rsid wildtype mutant  \\\n",
      "62  MC4R_001           -         -  g.57882787C>A/T  rs489693        C      A   \n",
      "\n",
      "    gene_id  marker_id  gene   marker_label  \n",
      "62       31         63  MC4R  MC4R rs489693  \n",
      "      marker star_allele aa_change        nt_change      rsid wildtype mutant  \\\n",
      "62  MC4R_001           -         -  g.57882787C>A/T  rs489693        C      A   \n",
      "\n",
      "    gene_id  marker_id  gene   marker_label  \n",
      "62       31         63  MC4R  MC4R rs489693  \n",
      "      marker star_allele aa_change        nt_change      rsid wildtype mutant  \\\n",
      "62  MC4R_001           -         -  g.57882787C>A/T  rs489693        C      A   \n",
      "\n",
      "    gene_id  marker_id  gene   marker_label  \n",
      "62       31         63  MC4R  MC4R rs489693  \n",
      "        marker star_allele        aa_change nt_change       rsid wildtype  \\\n",
      "30  CYP2D6_011          *4  Splicing_defect   1846G>A  rs3892097        G   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "30      A       14         31  CYP2D6    CYP2D6 *4  \n",
      "        marker star_allele        aa_change nt_change       rsid wildtype  \\\n",
      "30  CYP2D6_011          *4  Splicing_defect   1846G>A  rs3892097        G   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "30      A       14         31  CYP2D6    CYP2D6 *4  \n",
      "        marker star_allele       aa_change nt_change        rsid wildtype  \\\n",
      "31  CYP2D6_012          *3  259_frameshift  2549delA  rs35742686        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "31      G       14         32  CYP2D6    CYP2D6 *3  \n",
      "        marker star_allele       aa_change nt_change        rsid wildtype  \\\n",
      "31  CYP2D6_012          *3  259_frameshift  2549delA  rs35742686        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "31      G       14         32  CYP2D6    CYP2D6 *3  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "32  CYP2D6_013         *17     T107I   1023C>T  rs28371706        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "32       14         33  CYP2D6   CYP2D6 *17  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "32  CYP2D6_013         *17     T107I   1023C>T  rs28371706        C      T   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "32       14         33  CYP2D6   CYP2D6 *17  \n",
      "        marker star_allele aa_change         nt_change       rsid wildtype  \\\n",
      "33  CYP2D6_014          *9   K281del  2615_2617_delAAG  rs5030656        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "33      G       14         34  CYP2D6    CYP2D6 *9  \n",
      "        marker star_allele aa_change         nt_change       rsid wildtype  \\\n",
      "33  CYP2D6_014          *9   K281del  2615_2617_delAAG  rs5030656        A   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "33      G       14         34  CYP2D6    CYP2D6 *9  \n",
      "        marker star_allele       aa_change nt_change       rsid wildtype  \\\n",
      "34  CYP2D6_015          *6  118_frameshift  1707delT  rs5030655        T   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "34      G       14         35  CYP2D6    CYP2D6 *6  \n",
      "        marker star_allele       aa_change nt_change       rsid wildtype  \\\n",
      "34  CYP2D6_015          *6  118_frameshift  1707delT  rs5030655        T   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "34      G       14         35  CYP2D6    CYP2D6 *6  \n",
      "        marker star_allele       aa_change nt_change       rsid wildtype  \\\n",
      "34  CYP2D6_015          *6  118_frameshift  1707delT  rs5030655        T   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "34      G       14         35  CYP2D6    CYP2D6 *6  \n",
      "        marker star_allele       aa_change nt_change       rsid wildtype  \\\n",
      "34  CYP2D6_015          *6  118_frameshift  1707delT  rs5030655        T   \n",
      "\n",
      "   mutant  gene_id  marker_id    gene marker_label  \n",
      "34      G       14         35  CYP2D6    CYP2D6 *6  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "35  CYP2D6_016         *29     V338M   3183G>A  rs59421388        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "35       14         36  CYP2D6   CYP2D6 *29  \n",
      "        marker star_allele aa_change nt_change        rsid wildtype mutant  \\\n",
      "35  CYP2D6_016         *29     V338M   3183G>A  rs59421388        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "35       14         36  CYP2D6   CYP2D6 *29  \n",
      "       marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "69  OPRM1_001           -      N40D  c.118A>G  rs1799971        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "69       36         70  OPRM1  OPRM1 rs1799971  \n",
      "       marker star_allele aa_change nt_change       rsid wildtype mutant  \\\n",
      "69  OPRM1_001           -      N40D  c.118A>G  rs1799971        A      G   \n",
      "\n",
      "    gene_id  marker_id   gene     marker_label  \n",
      "69       36         70  OPRM1  OPRM1 rs1799971  \n",
      "        marker star_allele aa_change    nt_change rsid wildtype mutant  \\\n",
      "36  CYP2D6_017         *XN         -  Duplication    -        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "36       14         37  CYP2D6   CYP2D6 *XN  \n",
      "        marker star_allele aa_change    nt_change rsid wildtype mutant  \\\n",
      "36  CYP2D6_017         *XN         -  Duplication    -        G      A   \n",
      "\n",
      "    gene_id  marker_id    gene marker_label  \n",
      "36       14         37  CYP2D6   CYP2D6 *XN  \n"
     ]
    }
   ],
   "source": [
    "bin_file = './validation_panels/panels/[2021.08 └┌╖ß] PGx-NP_PGx-NP_bins_lam_updated.txt'\n",
    "bin = BinFile(bin_file)\n",
    "bin.process_data.to_excel('./validation_panels/PGx_NP_bin_seting_table_new_form.xlsx', index=False)\n",
    "data = bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>version</th>\n",
       "      <th>defined_panel</th>\n",
       "      <th>panel</th>\n",
       "      <th>gene</th>\n",
       "      <th>marker</th>\n",
       "      <th>marker_label</th>\n",
       "      <th>is_forward</th>\n",
       "      <th>direction</th>\n",
       "      <th>base</th>\n",
       "      <th>basetype</th>\n",
       "      <th>color</th>\n",
       "      <th>min_height</th>\n",
       "      <th>min_bin</th>\n",
       "      <th>max_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>CYP2D6 *49</td>\n",
       "      <td>0</td>\n",
       "      <td>Reverse</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>37.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>UGT1A4_001</td>\n",
       "      <td>UGT1A4 142T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>63.5</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>UGT1A4_001</td>\n",
       "      <td>UGT1A4 142T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>mutant</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>61.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT2B15</td>\n",
       "      <td>UGT2B15_001</td>\n",
       "      <td>UGT2B15 253G&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>51.8</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT2B15</td>\n",
       "      <td>UGT2B15_001</td>\n",
       "      <td>UGT2B15 253G&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>52.0</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    package   version defined_panel panel     gene       marker  \\\n",
       "1    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_001   \n",
       "0    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_001   \n",
       "2    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_002   \n",
       "3    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_002   \n",
       "5    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_003   \n",
       "..      ...       ...           ...   ...      ...          ...   \n",
       "107  PGx-NP  GM v 4.1       N-SPM06    S6     MC4R     MC4R_001   \n",
       "109  PGx-NP  GM v 4.1       N-SPM06    S6   UGT1A4   UGT1A4_001   \n",
       "108  PGx-NP  GM v 4.1       N-SPM06    S6   UGT1A4   UGT1A4_001   \n",
       "110  PGx-NP  GM v 4.1       N-SPM06    S6  UGT2B15  UGT2B15_001   \n",
       "111  PGx-NP  GM v 4.1       N-SPM06    S6  UGT2B15  UGT2B15_001   \n",
       "\n",
       "       marker_label  is_forward direction base  basetype  color  min_height  \\\n",
       "1        CYP2D6 *14           1   Forward    G  wildtype   blue        1000   \n",
       "0        CYP2D6 *14           1   Forward    A    mutant  green        1000   \n",
       "2       CYP2D6 *10B           1   Forward    C  wildtype  black        1000   \n",
       "3       CYP2D6 *10B           1   Forward    T    mutant    red        1000   \n",
       "5        CYP2D6 *49           0   Reverse    T  wildtype  green        1000   \n",
       "..              ...         ...       ...  ...       ...    ...         ...   \n",
       "107   MC4R rs489693           1   Forward    T    mutant    red        1000   \n",
       "109   UGT1A4 142T>G           1   Forward    T  wildtype    red        1000   \n",
       "108   UGT1A4 142T>G           1   Forward    G    mutant   blue        1000   \n",
       "110  UGT2B15 253G>T           1   Forward    G  wildtype   blue        1000   \n",
       "111  UGT2B15 253G>T           1   Forward    T    mutant    red        1000   \n",
       "\n",
       "     min_bin  max_bin  \n",
       "1       25.0     35.0  \n",
       "0       27.0     36.0  \n",
       "2       28.0     38.0  \n",
       "3       31.0     37.0  \n",
       "5       37.0     44.0  \n",
       "..       ...      ...  \n",
       "107     76.0     84.0  \n",
       "109     63.5     69.0  \n",
       "108     61.0     67.0  \n",
       "110     51.8     57.0  \n",
       "111     52.0     58.0  \n",
       "\n",
       "[112 rows x 15 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = data.groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).agg(bin_min=('min_bin', 'min'), bin_max=('max_bin', 'max'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>panel</th>\n",
       "      <th>marker</th>\n",
       "      <th>base</th>\n",
       "      <th>color</th>\n",
       "      <th>bin_min</th>\n",
       "      <th>bin_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>A</td>\n",
       "      <td>green</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>G</td>\n",
       "      <td>blue</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>C</td>\n",
       "      <td>black</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>T</td>\n",
       "      <td>red</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>A</td>\n",
       "      <td>red</td>\n",
       "      <td>39.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    version panel      marker base  color  bin_min  bin_max\n",
       "0  GM v 4.1    S1  CYP2D6_001    A  green     27.0     36.0\n",
       "1  GM v 4.1    S1  CYP2D6_001    G   blue     25.0     35.0\n",
       "2  GM v 4.1    S1  CYP2D6_002    C  black     28.0     38.0\n",
       "3  GM v 4.1    S1  CYP2D6_002    T    red     31.0     37.0\n",
       "4  GM v 4.1    S1  CYP2D6_003    A    red     39.0     45.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>version</th>\n",
       "      <th>defined_panel</th>\n",
       "      <th>panel</th>\n",
       "      <th>gene</th>\n",
       "      <th>marker</th>\n",
       "      <th>marker_label</th>\n",
       "      <th>is_forward</th>\n",
       "      <th>direction</th>\n",
       "      <th>base</th>\n",
       "      <th>basetype</th>\n",
       "      <th>color</th>\n",
       "      <th>min_height</th>\n",
       "      <th>min_bin</th>\n",
       "      <th>max_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>CYP2D6 *49</td>\n",
       "      <td>0</td>\n",
       "      <td>Reverse</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>37.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.5</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    package   version defined_panel panel    gene      marker   marker_label  \\\n",
       "1    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "0    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "2    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "3    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "5    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_003     CYP2D6 *49   \n",
       "..      ...       ...           ...   ...     ...         ...            ...   \n",
       "92   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "93   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "106  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "105  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "107  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "\n",
       "     is_forward direction base  basetype  color  min_height  min_bin  max_bin  \n",
       "1             1   Forward    G  wildtype   blue        1000     25.0     35.0  \n",
       "0             1   Forward    A    mutant  green        1000     27.0     36.0  \n",
       "2             1   Forward    C  wildtype  black        1000     28.0     38.0  \n",
       "3             1   Forward    T    mutant    red        1000     31.0     37.0  \n",
       "5             0   Reverse    T  wildtype  green        1000     37.0     44.0  \n",
       "..          ...       ...  ...       ...    ...         ...      ...      ...  \n",
       "92            1   Forward    A    mutant  green        1000     68.0     75.0  \n",
       "93            1   Forward    C  wildtype  black        1000     68.5     75.0  \n",
       "106           1   Forward    C  wildtype  black        1000     76.0     84.0  \n",
       "105           1   Forward    A    mutant  green        1000     76.0     83.0  \n",
       "107           1   Forward    T    mutant    red        1000     76.0     84.0  \n",
       "\n",
       "[112 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from Source import generate_markers\n",
    "import pandas as pd\n",
    "data = pd.read_excel(\"validation_panels/PGx_NP_bin_seting_table_new_form.xlsx\")\n",
    "markers = generate_markers(data, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].defined_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].color"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('snapshot_genotyping')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b4d3fe072432f0816511cf4002824cbf67c2cdbd34bdb0afb8cba97d824c69ea"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
