{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Source import data_chanel_map\n",
    "from Source import forward_color\n",
    "from Source import reverse_color\n",
    "from Source import reverse_map\n",
    "from Source import Definition\n",
    "\n",
    "definition = Definition(\"./resource/tables.pdata\")\n",
    "\n",
    "class BinFile:\n",
    "    def __init__(self, file) -> None:\n",
    "        self.data = self.read_file(file)\n",
    "        self.process_data = self.parser()\n",
    "\n",
    "    def read_file(self, file):\n",
    "        with open(file, 'r') as f:\n",
    "            return [line.strip('\\n') for line in f.readlines()]\n",
    "\n",
    "    def parser(self):\n",
    "        lines = []\n",
    "        for line in self.data:\n",
    "            # print(line)\n",
    "            if 'Version' in line:\n",
    "                version = line.split('\\t')[1]\n",
    "            elif 'Chemistry' in line:\n",
    "                kit_name = line.split('\\t')[1]\n",
    "            elif 'Panel Name' in line:\n",
    "                panel_name = line.split('\\t')[1]\n",
    "            elif 'Marker Name' in line:\n",
    "                marker_name = line.split('\\t')[1]\n",
    "            elif 'BinSet' in line:\n",
    "                continue\n",
    "            else:\n",
    "                # print(panel_name)\n",
    "                line_base = line.split('\\t')\n",
    "                line_base.insert(0, marker_name)\n",
    "                line_base.insert(0, panel_name)\n",
    "                line_base.insert(0, kit_name)\n",
    "                line_base.insert(0, version)\n",
    "                lines.append(line_base)\n",
    "\n",
    "        \n",
    "        df = pd.DataFrame(lines, columns=['version', 'package', 'defined_panel', 'marker', 'base', 'min_bin', 'max_bin', 'color'])\n",
    "        df = df[pd.notnull(df.color)]\n",
    "        df['min_bin'] = round(df['min_bin'].astype('float'),2)\n",
    "        df['max_bin'] = round(df['max_bin'].astype('float'),2)\n",
    "        df['color'] = df['color'].apply(lambda x: x.lower())\n",
    "        \n",
    "        # conver yellow to black\n",
    "        df['color'] = df['color'].apply(lambda x: 'black' if x  == 'yellow' else x)\n",
    "        \n",
    "        # update direction\n",
    "        df['direction'] = df.apply(lambda row: self.update_direction(row['base'], row['color']), axis=1)\n",
    "\n",
    "        # update is_forward\n",
    "        df['is_forward'] = df['direction'].apply(lambda x: 1 if x == 'Forward' else 0)\n",
    "\n",
    "        # update label\n",
    "        df = df.merge(definition.marker_table[['marker', 'marker_label', 'gene']], how='left')\n",
    "\n",
    "        # update panel\n",
    "        df['panel'] = df['defined_panel'].apply(lambda x: f'S{x[-1]}')\n",
    "\n",
    "        # update basetype\n",
    "\n",
    "        def _base_type(marker, base):\n",
    "            # print(marker)\n",
    "            marker_data = definition.marker_table[definition.marker_table['marker'] ==  marker]\n",
    "            print(marker_data)\n",
    "            wildtype = marker_data['wildtype'].to_list()[0]\n",
    "\n",
    "            return 'wildtype' if base == wildtype else 'mutant'\n",
    "\n",
    "\n",
    "        df['basetype'] = df.apply(lambda row: _base_type(row['marker'], row['base']), axis=1)\n",
    "                \n",
    "\n",
    "        # intiate min intensity\n",
    "        df['min_height'] = 1000\n",
    "\n",
    "        # process duplicate\n",
    "        # get max range file of bin setting if base duplicated\n",
    "        # M1 20  25\n",
    "        # M1 21  26\n",
    "        # --> M1 20 26 \n",
    "\n",
    "        tmp = df.groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).agg(bin_min=('min_bin', 'min'), bin_max=('max_bin', 'max'))\n",
    "\n",
    "        df = df.merge(tmp)\n",
    "\n",
    "        df.drop(columns=['min_bin', 'max_bin'], inplace=True)\n",
    "        df.rename(columns={'bin_min':'min_bin', 'bin_max':'max_bin'}, inplace=True)\n",
    "\n",
    "        # commennt the above code if we want to get the max bin range instead\n",
    "\n",
    "        df['binrange'] = df.min_bin - df.max_bin\n",
    "\n",
    "        # get max bin range\n",
    "        df = df.sort_values('binrange', ascending=False).groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).first()\n",
    "\n",
    "        df.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True], inplace=True)\n",
    "       \n",
    "        # reorder colums\n",
    "        df = df[['package', 'version', 'defined_panel', 'panel', 'gene', 'marker',  'marker_label', 'is_forward',  'direction', \n",
    "                'base', 'basetype', 'color',  'min_height', 'min_bin', 'max_bin']]\n",
    "\n",
    "\n",
    "        # check status marker\n",
    "\n",
    "        markers = df[['panel','marker', 'direction']].drop_duplicates()\n",
    "        count = markers.groupby('marker').size()\n",
    "\n",
    "        if any(count[count > 1]):\n",
    "            error_markers = \"; \".join(count[count > 1].index)\n",
    "            raise ValueError(f'Marker(s) info not consistent [{error_markers}]')\n",
    "        \n",
    "        else:\n",
    "            return df\n",
    "    \n",
    "    def update_direction(self, base, color):\n",
    "        return 'Forward' if forward_color.get(base) == color else 'Reverse'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYP2D6_001\n",
      "CYP2D6_001\n",
      "CYP2D6_002\n",
      "CYP2D6_002\n",
      "CYP2D6_003\n",
      "CYP2D6_003\n",
      "CYP2D6_004\n",
      "CYP2D6_004\n",
      "CYP2D6_005\n",
      "CYP2D6_005\n",
      "CYP2D6_006\n",
      "CYP2D6_006\n",
      "CYP2D6_007\n",
      "CYP2D6_007\n",
      "CYP2D6_008\n",
      "CYP2D6_008\n",
      "CYP2D6_009\n",
      "CYP2D6_009\n",
      "CYP2D6_010\n",
      "CYP2D6_010\n",
      "CYP2D6_011\n",
      "CYP2D6_011\n",
      "VKORC1_001\n",
      "VKORC1_001\n",
      "CYP4F2_001\n",
      "CYP4F2_001\n",
      "CYP2C19_001\n",
      "CYP2C19_001\n",
      "CYP2C9_001\n",
      "CYP2C9_001\n",
      "CYP2C19_002\n",
      "CYP2C19_002\n",
      "CYP2C9_002\n",
      "CYP2C9_002\n",
      "CYP2C9_003\n",
      "CYP2C9_003\n",
      "CYP2C19_003\n",
      "CYP2C19_003\n",
      "CYP2C9_004\n",
      "CYP2C9_004\n",
      "CYP3A4_002\n",
      "CYP3A4_002\n",
      "CYP3A4_001\n",
      "CYP3A4_001\n",
      "CYP3A5_001\n",
      "CYP3A5_001\n",
      "C11orf65_001\n",
      "C11orf65_001\n",
      "SLCO1B1_001\n",
      "SLCO1B1_001\n",
      "MMP3_001\n",
      "MMP3_001\n",
      "SLCO1B1_002\n",
      "SLCO1B1_002\n",
      "ABCG2_001\n",
      "ABCG2_001\n",
      "MT-RNR1_001\n",
      "MT-RNR1_001\n",
      "COQ2_001\n",
      "COQ2_001\n",
      "KCNJ11_001\n",
      "KCNJ11_001\n",
      "NEDD4L_001\n",
      "NEDD4L_001\n",
      "GNB3_001\n",
      "GNB3_001\n",
      "NUDT15_002\n",
      "NUDT15_002\n",
      "NUDT15_003\n",
      "NUDT15_003\n",
      "TPMT_002\n",
      "TPMT_002\n",
      "TPMT_001\n",
      "TPMT_001\n",
      "G6PD_001\n",
      "G6PD_001\n",
      "G6PD_002\n",
      "G6PD_002\n",
      "NUDT15_001\n",
      "NUDT15_001\n",
      "DPYD_002\n",
      "DPYD_002\n",
      "DPYD_001\n",
      "DPYD_001\n",
      "CYP2D6_017\n",
      "CYP2D6_017\n",
      "MDR1_002\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb Cell 2\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m bin_file \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m./validation_panels/panels/[2021.08 └┌╖ß] PGx-Premium_PGx-Premium_bins.txt\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mbin\u001b[39m \u001b[39m=\u001b[39m BinFile(bin_file)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m# bin.process_data.to_excel('./validation_panels/PGx_NP_bin_seting_table_new_form.xlsx', index=False)\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m data \u001b[39m=\u001b[39m \u001b[39mbin\u001b[39m\u001b[39m.\u001b[39mprocess_data\n",
      "\u001b[1;32m/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb Cell 2\u001b[0m in \u001b[0;36mBinFile.__init__\u001b[0;34m(self, file)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, file) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdata \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mread_file(file)\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprocess_data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mparser()\n",
      "\u001b[1;32m/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb Cell 2\u001b[0m in \u001b[0;36mBinFile.parser\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=68'>69</a>\u001b[0m     wildtype \u001b[39m=\u001b[39m marker_data[\u001b[39m'\u001b[39m\u001b[39mwildtype\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mto_list()[\u001b[39m0\u001b[39m]\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=70'>71</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mwildtype\u001b[39m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m base \u001b[39m==\u001b[39m wildtype \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mmutant\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=73'>74</a>\u001b[0m df[\u001b[39m'\u001b[39m\u001b[39mbasetype\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m df\u001b[39m.\u001b[39;49mapply(\u001b[39mlambda\u001b[39;49;00m row: _base_type(row[\u001b[39m'\u001b[39;49m\u001b[39mmarker\u001b[39;49m\u001b[39m'\u001b[39;49m], row[\u001b[39m'\u001b[39;49m\u001b[39mbase\u001b[39;49m\u001b[39m'\u001b[39;49m]), axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=76'>77</a>\u001b[0m \u001b[39m# intiate min intensity\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=77'>78</a>\u001b[0m df[\u001b[39m'\u001b[39m\u001b[39mmin_height\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1000\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/snapshot_genotyping/lib/python3.8/site-packages/pandas/core/frame.py:8848\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[1;32m   8837\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpandas\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcore\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mapply\u001b[39;00m \u001b[39mimport\u001b[39;00m frame_apply\n\u001b[1;32m   8839\u001b[0m op \u001b[39m=\u001b[39m frame_apply(\n\u001b[1;32m   8840\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m   8841\u001b[0m     func\u001b[39m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   8846\u001b[0m     kwargs\u001b[39m=\u001b[39mkwargs,\n\u001b[1;32m   8847\u001b[0m )\n\u001b[0;32m-> 8848\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mapply()\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mapply\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/snapshot_genotyping/lib/python3.8/site-packages/pandas/core/apply.py:733\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    730\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw:\n\u001b[1;32m    731\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_raw()\n\u001b[0;32m--> 733\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[0;32m~/miniconda3/envs/snapshot_genotyping/lib/python3.8/site-packages/pandas/core/apply.py:857\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    856\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply_standard\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 857\u001b[0m     results, res_index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_series_generator()\n\u001b[1;32m    859\u001b[0m     \u001b[39m# wrap results\u001b[39;00m\n\u001b[1;32m    860\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m~/miniconda3/envs/snapshot_genotyping/lib/python3.8/site-packages/pandas/core/apply.py:873\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    870\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    871\u001b[0m     \u001b[39mfor\u001b[39;00m i, v \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(series_gen):\n\u001b[1;32m    872\u001b[0m         \u001b[39m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         results[i] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mf(v)\n\u001b[1;32m    874\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    875\u001b[0m             \u001b[39m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    876\u001b[0m             \u001b[39m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    877\u001b[0m             results[i] \u001b[39m=\u001b[39m results[i]\u001b[39m.\u001b[39mcopy(deep\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "\u001b[1;32m/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb Cell 2\u001b[0m in \u001b[0;36mBinFile.parser.<locals>.<lambda>\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=68'>69</a>\u001b[0m     wildtype \u001b[39m=\u001b[39m marker_data[\u001b[39m'\u001b[39m\u001b[39mwildtype\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mto_list()[\u001b[39m0\u001b[39m]\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=70'>71</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mwildtype\u001b[39m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m base \u001b[39m==\u001b[39m wildtype \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mmutant\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=73'>74</a>\u001b[0m df[\u001b[39m'\u001b[39m\u001b[39mbasetype\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m df\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m row: _base_type(row[\u001b[39m'\u001b[39;49m\u001b[39mmarker\u001b[39;49m\u001b[39m'\u001b[39;49m], row[\u001b[39m'\u001b[39;49m\u001b[39mbase\u001b[39;49m\u001b[39m'\u001b[39;49m]), axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=76'>77</a>\u001b[0m \u001b[39m# intiate min intensity\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=77'>78</a>\u001b[0m df[\u001b[39m'\u001b[39m\u001b[39mmin_height\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1000\u001b[39m\n",
      "\u001b[1;32m/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb Cell 2\u001b[0m in \u001b[0;36mBinFile.parser.<locals>._base_type\u001b[0;34m(marker, base)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=66'>67</a>\u001b[0m \u001b[39mprint\u001b[39m(marker)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=67'>68</a>\u001b[0m marker_data \u001b[39m=\u001b[39m definition\u001b[39m.\u001b[39mmarker_table[definition\u001b[39m.\u001b[39mmarker_table[\u001b[39m'\u001b[39m\u001b[39mmarker\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m  marker]\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=68'>69</a>\u001b[0m wildtype \u001b[39m=\u001b[39m marker_data[\u001b[39m'\u001b[39;49m\u001b[39mwildtype\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m.\u001b[39;49mto_list()[\u001b[39m0\u001b[39;49m]\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/lamnguyen/Works/snapshot_genotyping/process_bin_setting.ipynb#W1sZmlsZQ%3D%3D?line=70'>71</a>\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mwildtype\u001b[39m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m base \u001b[39m==\u001b[39m wildtype \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mmutant\u001b[39m\u001b[39m'\u001b[39m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "bin_file = './validation_panels/panels/[2021.08 └┌╖ß] PGx-Premium_PGx-Premium_bins.txt'\n",
    "bin = BinFile(bin_file)\n",
    "# bin.process_data.to_excel('./validation_panels/PGx_NP_bin_seting_table_new_form.xlsx', index=False)\n",
    "data = bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>version</th>\n",
       "      <th>defined_panel</th>\n",
       "      <th>panel</th>\n",
       "      <th>gene</th>\n",
       "      <th>marker</th>\n",
       "      <th>marker_label</th>\n",
       "      <th>is_forward</th>\n",
       "      <th>direction</th>\n",
       "      <th>base</th>\n",
       "      <th>basetype</th>\n",
       "      <th>color</th>\n",
       "      <th>min_height</th>\n",
       "      <th>min_bin</th>\n",
       "      <th>max_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>CYP2D6 *49</td>\n",
       "      <td>0</td>\n",
       "      <td>Reverse</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>37.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>UGT1A4_001</td>\n",
       "      <td>UGT1A4 142T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>63.5</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>UGT1A4_001</td>\n",
       "      <td>UGT1A4 142T&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>mutant</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>61.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT2B15</td>\n",
       "      <td>UGT2B15_001</td>\n",
       "      <td>UGT2B15 253G&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>51.8</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>UGT2B15</td>\n",
       "      <td>UGT2B15_001</td>\n",
       "      <td>UGT2B15 253G&gt;T</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>52.0</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    package   version defined_panel panel     gene       marker  \\\n",
       "1    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_001   \n",
       "0    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_001   \n",
       "2    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_002   \n",
       "3    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_002   \n",
       "5    PGx-NP  GM v 4.1       N-SPM01    S1   CYP2D6   CYP2D6_003   \n",
       "..      ...       ...           ...   ...      ...          ...   \n",
       "107  PGx-NP  GM v 4.1       N-SPM06    S6     MC4R     MC4R_001   \n",
       "109  PGx-NP  GM v 4.1       N-SPM06    S6   UGT1A4   UGT1A4_001   \n",
       "108  PGx-NP  GM v 4.1       N-SPM06    S6   UGT1A4   UGT1A4_001   \n",
       "110  PGx-NP  GM v 4.1       N-SPM06    S6  UGT2B15  UGT2B15_001   \n",
       "111  PGx-NP  GM v 4.1       N-SPM06    S6  UGT2B15  UGT2B15_001   \n",
       "\n",
       "       marker_label  is_forward direction base  basetype  color  min_height  \\\n",
       "1        CYP2D6 *14           1   Forward    G  wildtype   blue        1000   \n",
       "0        CYP2D6 *14           1   Forward    A    mutant  green        1000   \n",
       "2       CYP2D6 *10B           1   Forward    C  wildtype  black        1000   \n",
       "3       CYP2D6 *10B           1   Forward    T    mutant    red        1000   \n",
       "5        CYP2D6 *49           0   Reverse    T  wildtype  green        1000   \n",
       "..              ...         ...       ...  ...       ...    ...         ...   \n",
       "107   MC4R rs489693           1   Forward    T    mutant    red        1000   \n",
       "109   UGT1A4 142T>G           1   Forward    T  wildtype    red        1000   \n",
       "108   UGT1A4 142T>G           1   Forward    G    mutant   blue        1000   \n",
       "110  UGT2B15 253G>T           1   Forward    G  wildtype   blue        1000   \n",
       "111  UGT2B15 253G>T           1   Forward    T    mutant    red        1000   \n",
       "\n",
       "     min_bin  max_bin  \n",
       "1       25.0     35.0  \n",
       "0       27.0     36.0  \n",
       "2       28.0     38.0  \n",
       "3       31.0     37.0  \n",
       "5       37.0     44.0  \n",
       "..       ...      ...  \n",
       "107     76.0     84.0  \n",
       "109     63.5     69.0  \n",
       "108     61.0     67.0  \n",
       "110     51.8     57.0  \n",
       "111     52.0     58.0  \n",
       "\n",
       "[112 rows x 15 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = data.groupby(['version', 'panel', 'marker', 'base', 'color'], as_index=False).agg(bin_min=('min_bin', 'min'), bin_max=('max_bin', 'max'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>panel</th>\n",
       "      <th>marker</th>\n",
       "      <th>base</th>\n",
       "      <th>color</th>\n",
       "      <th>bin_min</th>\n",
       "      <th>bin_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>A</td>\n",
       "      <td>green</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>G</td>\n",
       "      <td>blue</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>C</td>\n",
       "      <td>black</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>T</td>\n",
       "      <td>red</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>A</td>\n",
       "      <td>red</td>\n",
       "      <td>39.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    version panel      marker base  color  bin_min  bin_max\n",
       "0  GM v 4.1    S1  CYP2D6_001    A  green     27.0     36.0\n",
       "1  GM v 4.1    S1  CYP2D6_001    G   blue     25.0     35.0\n",
       "2  GM v 4.1    S1  CYP2D6_002    C  black     28.0     38.0\n",
       "3  GM v 4.1    S1  CYP2D6_002    T    red     31.0     37.0\n",
       "4  GM v 4.1    S1  CYP2D6_003    A    red     39.0     45.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.sort_values(['panel', 'gene', 'marker', 'basetype', 'min_bin'], ascending=[True, True, True, False, True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>version</th>\n",
       "      <th>defined_panel</th>\n",
       "      <th>panel</th>\n",
       "      <th>gene</th>\n",
       "      <th>marker</th>\n",
       "      <th>marker_label</th>\n",
       "      <th>is_forward</th>\n",
       "      <th>direction</th>\n",
       "      <th>base</th>\n",
       "      <th>basetype</th>\n",
       "      <th>color</th>\n",
       "      <th>min_height</th>\n",
       "      <th>min_bin</th>\n",
       "      <th>max_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>G</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>blue</td>\n",
       "      <td>1000</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_001</td>\n",
       "      <td>CYP2D6 *14</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_002</td>\n",
       "      <td>CYP2D6 *10B</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM01</td>\n",
       "      <td>S1</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6_003</td>\n",
       "      <td>CYP2D6 *49</td>\n",
       "      <td>0</td>\n",
       "      <td>Reverse</td>\n",
       "      <td>T</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>37.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>CYP1A2_001</td>\n",
       "      <td>CYP1A2 *1F</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>68.5</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>C</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>black</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>A</td>\n",
       "      <td>mutant</td>\n",
       "      <td>green</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PGx-NP</td>\n",
       "      <td>GM v 4.1</td>\n",
       "      <td>N-SPM06</td>\n",
       "      <td>S6</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>MC4R_001</td>\n",
       "      <td>MC4R rs489693</td>\n",
       "      <td>1</td>\n",
       "      <td>Forward</td>\n",
       "      <td>T</td>\n",
       "      <td>mutant</td>\n",
       "      <td>red</td>\n",
       "      <td>1000</td>\n",
       "      <td>76.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    package   version defined_panel panel    gene      marker   marker_label  \\\n",
       "1    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "0    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_001     CYP2D6 *14   \n",
       "2    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "3    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_002    CYP2D6 *10B   \n",
       "5    PGx-NP  GM v 4.1       N-SPM01    S1  CYP2D6  CYP2D6_003     CYP2D6 *49   \n",
       "..      ...       ...           ...   ...     ...         ...            ...   \n",
       "92   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "93   PGx-NP  GM v 4.1       N-SPM06    S6  CYP1A2  CYP1A2_001     CYP1A2 *1F   \n",
       "106  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "105  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "107  PGx-NP  GM v 4.1       N-SPM06    S6    MC4R    MC4R_001  MC4R rs489693   \n",
       "\n",
       "     is_forward direction base  basetype  color  min_height  min_bin  max_bin  \n",
       "1             1   Forward    G  wildtype   blue        1000     25.0     35.0  \n",
       "0             1   Forward    A    mutant  green        1000     27.0     36.0  \n",
       "2             1   Forward    C  wildtype  black        1000     28.0     38.0  \n",
       "3             1   Forward    T    mutant    red        1000     31.0     37.0  \n",
       "5             0   Reverse    T  wildtype  green        1000     37.0     44.0  \n",
       "..          ...       ...  ...       ...    ...         ...      ...      ...  \n",
       "92            1   Forward    A    mutant  green        1000     68.0     75.0  \n",
       "93            1   Forward    C  wildtype  black        1000     68.5     75.0  \n",
       "106           1   Forward    C  wildtype  black        1000     76.0     84.0  \n",
       "105           1   Forward    A    mutant  green        1000     76.0     83.0  \n",
       "107           1   Forward    T    mutant    red        1000     76.0     84.0  \n",
       "\n",
       "[112 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin.process_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from Source import generate_markers\n",
    "import pandas as pd\n",
    "data = pd.read_excel(\"validation_panels/PGx_NP_bin_seting_table_new_form.xlsx\")\n",
    "markers = generate_markers(data, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].defined_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'green'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers['CYP2D6_001'].alleles[0].color"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('snapshot_genotyping')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b4d3fe072432f0816511cf4002824cbf67c2cdbd34bdb0afb8cba97d824c69ea"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
