<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Customized star allele calling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="star_allele_calling_files/libs/clipboard/clipboard.min.js"></script>
<script src="star_allele_calling_files/libs/quarto-html/quarto.js"></script>
<script src="star_allele_calling_files/libs/quarto-html/popper.min.js"></script>
<script src="star_allele_calling_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="star_allele_calling_files/libs/quarto-html/anchor.min.js"></script>
<link href="star_allele_calling_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="star_allele_calling_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="star_allele_calling_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="star_allele_calling_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="star_allele_calling_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Customized star allele calling</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="introduction" class="level1">
<h1>Introduction</h1>
<ul>
<li>Instroduction about SNAPShot</li>
<li>The process of genotype calling</li>
<li>Vary each run time</li>
<li>Benefit of cloud base system</li>
</ul>
<p>Star-allele calling from genotyping results is crucial for PGx implementation in clinical practice. Additionally, a laboratory-developed test (LDT) panel for PGx implementation is also critical and cost-effective for a specific population; however, the star-alleles calling tools for the panel need to be developed. In this work, therefore, we developed a star-allele calling tool applied for the developed testing panel. The tool also was integrated into the system and website application.</p>
<p>Taken all into consideration, we develop a cloud base system that support user can call genotype and matching with our knowledge database.</p>
</section>
<section id="methods" class="level1">
<h1>Methods</h1>
<section id="loading-fsa-data" class="level2">
<h2 class="anchored" data-anchor-id="loading-fsa-data">Loading FSA data</h2>
<p>Raw data from fragment analysis sofware were loaded using package Biopython (version) package. The intensisty data were stored in different predefined chanel name such as DATA105 for refrernce chanel, for other chanels defined for 4 nucleotites (DATA1: A; DATA2:C; DATA3:G and DATA4:T). Those intensisty data were primary used in peak detection process.</p>
</section>
<section id="reference-peak-detection" class="level2">
<h2 class="anchored" data-anchor-id="reference-peak-detection">Reference peak detection</h2>
<p>As mentioned aboved, the DATA105 chanel as refrence intensity was used to detect the reference peak. Usually, the information of reference peak depends on the experimen desgin. In this study, we used ??? GeneScan 120 Lize dye size standard (CYP2D6 kit developed by SPMED Co.&nbsp;Ltd.&nbsp;(GTR link)).</p>
<p>We used peak finding function of scipy (version) with default height 800, width of peak, â€¦ However, the setting information can be adjusted by experties.</p>
<p>The number of detected peaks should be equal to number of reference size standard. In case of LIZ 120, there are 9 sizes, therefore the number of peaks should be 9 peaks. Only qualifed peak detection was used to develop Least Square model as the next step, otherwise, it need to be adjusted by user to detect the correct peaks.</p>
</section>
<section id="sizing-model-development" class="level2">
<h2 class="anchored" data-anchor-id="sizing-model-development">Sizing model development</h2>
<p>As described in the manual of gene mapper software, there were three method to interfer length of DNA based on intensity points (Local Southern, Global Southern and Least Square). In the work, we used Least Square as advance method to develop a model for sizing identification.</p>
<p>The model was developed using numpy (version) python package. Users can choose the order of least sequare method (second-order or third-order). The performance of this model was evaluated using R square index from actual size and precited size. Based on R square, we found that third-order resulted in a slightly higher than second order, therefore, thrid order was set as default of the model. The model development process will be done by each input FSA file.</p>
<p>Workflow</p>
<p>Please see here: https://app.diagrams.net/#G1HzMMyRkVhr8Y_M1wMmXwmZagIEpak1Bb</p>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> r2_score</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.signal <span class="im">import</span> find_peaks, peak_prominences, peak_widths</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">'ignore'</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> loadFSA(aFSA):</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Load FSA file</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">        aFSA (str): A path to a FSA file.</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co">        SeqIO: A data loaded from FSA file.</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(SeqIO.read(aFSA, <span class="st">'abi'</span>))</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generateReferenceRange(reference):</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Generate all range from reference data</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="co">        reference (list, required): A list contain all bp locations of refernces.</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co">                                    Defaults to [15, 20, 25, 35, 50, 62, 80, 110, 120].</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="co">        list:   The list contant all generated range of the references</span></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="co">                Example: [(15, 20), (20,25)]</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>    referenceRange <span class="op">=</span> []</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i,_ <span class="kw">in</span> <span class="bu">enumerate</span>(reference):</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> i <span class="op">&lt;</span> <span class="bu">len</span>(reference)<span class="op">-</span><span class="dv">1</span>:</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>            referenceRange.append((reference[i], reference[i<span class="op">+</span><span class="dv">1</span>]))</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> referenceRange</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> getRawData(FSA, chanel):</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Extract raw intensity by chanel.</span></span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a><span class="co">        FSA (class): A class load FSA file.</span></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a><span class="co">        chanel (str): An interested intensity chanel.</span></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a><span class="co">        Tuple: A tuple about a chanel intensity data.</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> FSA.rawdata[chanel]</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> FSA:</span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, aFSA) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.data <span class="op">=</span> SeqIO.read(aFSA, <span class="st">'abi'</span>)</span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.name <span class="op">=</span> <span class="va">self</span>.data.name</span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.<span class="bu">id</span> <span class="op">=</span> <span class="va">self</span>.data.<span class="bu">id</span></span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.description <span class="op">=</span> <span class="va">self</span>.data.description</span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.rawdata <span class="op">=</span> <span class="va">self</span>.data.annotations[<span class="st">'abif_raw'</span>]</span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> panel(<span class="va">self</span>):</span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Extract panel from fsa name. The first element of the name was consider as the panel.</span></span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a><span class="co">            str: name of the panel</span></span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.name.split(<span class="st">'-'</span>)[<span class="dv">0</span>]</span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> reference_intensity(<span class="va">self</span>):</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Extract reference intensity, in the fsa raw data format, DATA105 is defined for reference intensity chanel.</span></span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a><span class="co">            tupe: A tupe of all reference intensity data.</span></span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.rawdata[<span class="st">'DATA105'</span>]</span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> chanel_size(<span class="va">self</span>):</span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Return chanel size, it will be used for generate number of points and make basepairs prediction.</span></span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a><span class="co">            int: Length of the intensity chanel.</span></span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="bu">len</span>(<span class="va">self</span>.reference_intensity)</span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> extract_intensity(<span class="va">self</span>, chanel, <span class="bu">range</span><span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Extract the intensity by a specific chanel.</span></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a><span class="co">           In the FSA file, there 4 chanels (DATA1, DATA2, DATA3, DATA4) following 4 nucleotides (ATCG)</span></span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a><span class="co">            chanel (str): chanel name for [DATA1, DATA2, DATA3, DATA4]</span></span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a><span class="co">            range (tupe, optional): A tupe defines an intersted range for extract. It was resulted in by `get_index_by_base_range` function. Defaults to None.</span></span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a><span class="co">            tupe: intensity of the selected chanel</span></span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">range</span>:</span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="va">self</span>.rawdata[chanel][<span class="bu">range</span>[<span class="dv">0</span>]:<span class="bu">range</span>[<span class="dv">1</span>]]</span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="va">self</span>.rawdata[chanel]</span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> sample_name(<span class="va">self</span>):</span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> extract_sample_name(<span class="va">self</span>.name)</span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Reference:</span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, aRef) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.size <span class="op">=</span> aRef</span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.<span class="bu">range</span> <span class="op">=</span> generateReferenceRange(<span class="va">self</span>.size)</span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.length <span class="op">=</span> <span class="bu">len</span>(<span class="va">self</span>.size)</span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> finding_peaks(intensity, min_width<span class="op">=</span><span class="va">None</span>, min_height<span class="op">=</span><span class="dv">500</span>, min_prominence <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Finding Peaks based on intensity data</span></span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a><span class="co">        intensity (tupe, required): intensity data from a chanel</span></span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a><span class="co">        min_width (int, optional): threshold minimum width of the peaks. Defaults to None. Reference value: 1</span></span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a><span class="co">        min_height (int, optional): threshold minimum height of the peaks. Defaults to 500.</span></span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a><span class="co">        min_prominence (int, optional): threshold minimum depth of the peaks. Defaults to None. Reference value 50</span></span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a><span class="co">        peaks: a list of detected peaks</span></span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a><span class="co">        heights: a list of corresponding detected peaks</span></span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a>    peaks, heights <span class="op">=</span> find_peaks(intensity, height<span class="op">=</span>min_height)</span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a>    heights <span class="op">=</span> heights[<span class="st">'peak_heights'</span>]</span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> min_prominence:</span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a>        prominence <span class="op">=</span> peak_prominences(intensity, peaks)</span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a>        conditions <span class="op">=</span>  np.where(prominence[<span class="dv">0</span>] <span class="op">&gt;=</span> min_prominence)</span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a>        peaks <span class="op">=</span> peaks[conditions]</span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a>        heights <span class="op">=</span> heights[conditions]</span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> min_width:</span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a>        width <span class="op">=</span> peak_widths(intensity, peaks)</span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a>        conditions <span class="op">=</span> np.where(width[<span class="dv">0</span>] <span class="op">&gt;=</span> min_width)</span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a>        peaks <span class="op">=</span> peaks[conditions]</span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a>        heights <span class="op">=</span> heights[conditions]</span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> peaks, heights</span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> LeastSquared:</span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Building a LeastSequared model to predict the size of DNA based on intensity data points.</span></span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a><span class="co">       Each fsa can be different each run, therefore, model will be build based on specific data from FSA and reference.</span></span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a><span class="co">       The method of model was refer from gene mapper reference. More information can be found here:</span></span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a><span class="co">        Sizing: https://www2.unbc.ca/sites/default/files/sections/genetics/microsat.pdf</span></span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a><span class="co">        Peak scanner:  https://apps.thermofisher.com/apps/peak-scanner/help/GUID-CEE18138-6788-48FA-8E08-4CE939F4B9D8.html</span></span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, peak, reference_size, degree <span class="op">=</span> <span class="dv">3</span>):</span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.peak <span class="op">=</span> peak</span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.reference_size <span class="op">=</span> reference_size</span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.degree <span class="op">=</span> degree</span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.popt <span class="op">=</span> np.polyfit(<span class="va">self</span>.peak, <span class="va">self</span>.reference_size, <span class="va">self</span>.degree)</span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.ref_predicted <span class="op">=</span> np.polyval(<span class="va">self</span>.popt, <span class="va">self</span>.peak)</span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a>        <span class="co"># self.ref_predicted = np.round(self.ref_predicted, 2)</span></span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> r2_score(<span class="va">self</span>):</span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> r2_score(<span class="va">self</span>.reference_size, <span class="va">self</span>.ref_predicted)</span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> val_plot(<span class="va">self</span>, title):</span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Plot a validation figure from reference and predicted result.</span></span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a><span class="co">            title (str): Title of the figure, it can be a fsa name.</span></span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a>        _min <span class="op">=</span> np.<span class="bu">min</span>(<span class="va">self</span>.peak) <span class="op">-</span> <span class="dv">100</span></span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a>        _max <span class="op">=</span> np.<span class="bu">max</span>(<span class="va">self</span>.peak) <span class="op">+</span> <span class="dv">100</span></span>
<span id="cb1-190"><a href="#cb1-190" aria-hidden="true" tabindex="-1"></a>        xn <span class="op">=</span> <span class="bu">range</span>(_min, _max <span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb1-191"><a href="#cb1-191" aria-hidden="true" tabindex="-1"></a>        yn <span class="op">=</span> np.polyval(<span class="va">self</span>.popt, xn)</span>
<span id="cb1-192"><a href="#cb1-192" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-193"><a href="#cb1-193" aria-hidden="true" tabindex="-1"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">5</span>,<span class="dv">3</span>), dpi <span class="op">=</span> <span class="dv">150</span>)</span>
<span id="cb1-194"><a href="#cb1-194" aria-hidden="true" tabindex="-1"></a>        alpha <span class="op">=</span> <span class="fl">0.8</span></span>
<span id="cb1-195"><a href="#cb1-195" aria-hidden="true" tabindex="-1"></a>        ax.plot(xn, yn, color <span class="op">=</span> <span class="st">'blue'</span>, label <span class="op">=</span> <span class="st">'Predicted'</span>, alpha <span class="op">=</span> alpha)</span>
<span id="cb1-196"><a href="#cb1-196" aria-hidden="true" tabindex="-1"></a>        ax.plot(<span class="va">self</span>.peak, <span class="va">self</span>.reference_size, <span class="st">'or'</span>, label <span class="op">=</span> <span class="st">'Reference'</span>, alpha <span class="op">=</span>alpha)</span>
<span id="cb1-197"><a href="#cb1-197" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-198"><a href="#cb1-198" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> peak, size <span class="kw">in</span> <span class="bu">zip</span>(<span class="va">self</span>.peak, <span class="va">self</span>.reference_size):</span>
<span id="cb1-199"><a href="#cb1-199" aria-hidden="true" tabindex="-1"></a>            ax.text(peak, size <span class="op">+</span> <span class="dv">4</span>, <span class="ss">f'</span><span class="sc">{</span>size<span class="sc">}</span><span class="ss">'</span>, fontsize <span class="op">=</span><span class="dv">6</span>, ha<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb1-200"><a href="#cb1-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-201"><a href="#cb1-201" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Points'</span>)</span>
<span id="cb1-202"><a href="#cb1-202" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Size (bp)'</span>)</span>
<span id="cb1-203"><a href="#cb1-203" aria-hidden="true" tabindex="-1"></a>        ax.set_title(title, fontsize <span class="op">=</span> <span class="dv">8</span>)</span>
<span id="cb1-204"><a href="#cb1-204" aria-hidden="true" tabindex="-1"></a>        fig.text(<span class="fl">0.2</span>, <span class="fl">0.80</span>, <span class="ss">f'$R^</span><span class="sc">{</span><span class="dv">2</span><span class="sc">}</span><span class="ss"> = </span><span class="sc">{</span><span class="bu">str</span>(<span class="va">self</span>.r2_score)<span class="sc">}</span><span class="ss">$'</span>)</span>
<span id="cb1-205"><a href="#cb1-205" aria-hidden="true" tabindex="-1"></a>        plt.legend(loc<span class="op">=</span><span class="st">'lower right'</span>)</span>
<span id="cb1-206"><a href="#cb1-206" aria-hidden="true" tabindex="-1"></a>        plt.tight_layout()</span>
<span id="cb1-207"><a href="#cb1-207" aria-hidden="true" tabindex="-1"></a>        plt.show()</span>
<span id="cb1-208"><a href="#cb1-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-209"><a href="#cb1-209" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> predict(<span class="va">self</span>, x):</span>
<span id="cb1-210"><a href="#cb1-210" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Predict basepairs based on range of points</span></span>
<span id="cb1-211"><a href="#cb1-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-212"><a href="#cb1-212" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-213"><a href="#cb1-213" aria-hidden="true" tabindex="-1"></a><span class="co">            x (list): A list of points want to be predicted by built model</span></span>
<span id="cb1-214"><a href="#cb1-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-215"><a href="#cb1-215" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-216"><a href="#cb1-216" aria-hidden="true" tabindex="-1"></a><span class="co">            np.array: An array includes predicted basepairs.</span></span>
<span id="cb1-217"><a href="#cb1-217" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-218"><a href="#cb1-218" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> np.<span class="bu">round</span>(np.polyval(<span class="va">self</span>.popt, x),<span class="dv">2</span>)</span>
<span id="cb1-219"><a href="#cb1-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-220"><a href="#cb1-220" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> fit(<span class="va">self</span>, x:<span class="bu">list</span>):</span>
<span id="cb1-221"><a href="#cb1-221" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""predicted base bairs</span></span>
<span id="cb1-222"><a href="#cb1-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-223"><a href="#cb1-223" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-224"><a href="#cb1-224" aria-hidden="true" tabindex="-1"></a><span class="co">            x (list): data points</span></span>
<span id="cb1-225"><a href="#cb1-225" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-226"><a href="#cb1-226" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.predicted_base <span class="op">=</span> <span class="va">self</span>.predict(x)</span>
<span id="cb1-227"><a href="#cb1-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-228"><a href="#cb1-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-229"><a href="#cb1-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-230"><a href="#cb1-230" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Allele:</span>
<span id="cb1-231"><a href="#cb1-231" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, base:<span class="bu">str</span>, basetype:<span class="bu">str</span>, is_forward:<span class="bu">bool</span>, min_bin:<span class="bu">int</span>, max_bin:<span class="bu">int</span>, min_height:<span class="bu">int</span> <span class="op">=</span> <span class="dv">500</span>) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-232"><a href="#cb1-232" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Create allele class</span></span>
<span id="cb1-233"><a href="#cb1-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-234"><a href="#cb1-234" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-235"><a href="#cb1-235" aria-hidden="true" tabindex="-1"></a><span class="co">            base (str): tested nucleotide [A,T,C,G]</span></span>
<span id="cb1-236"><a href="#cb1-236" aria-hidden="true" tabindex="-1"></a><span class="co">            basetype (str): whether base is wildtype or mutant</span></span>
<span id="cb1-237"><a href="#cb1-237" aria-hidden="true" tabindex="-1"></a><span class="co">            min_bin (int): min bin set for peak detection</span></span>
<span id="cb1-238"><a href="#cb1-238" aria-hidden="true" tabindex="-1"></a><span class="co">            max_bin (int): max bin set for peak detection</span></span>
<span id="cb1-239"><a href="#cb1-239" aria-hidden="true" tabindex="-1"></a><span class="co">            min_height (int, optional): min height for peak detection. Defaults to 500.</span></span>
<span id="cb1-240"><a href="#cb1-240" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-241"><a href="#cb1-241" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.base <span class="op">=</span> base</span>
<span id="cb1-242"><a href="#cb1-242" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.basetype <span class="op">=</span> basetype</span>
<span id="cb1-243"><a href="#cb1-243" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.min_bin <span class="op">=</span> min_bin</span>
<span id="cb1-244"><a href="#cb1-244" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.max_bin <span class="op">=</span> max_bin</span>
<span id="cb1-245"><a href="#cb1-245" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.min_height <span class="op">=</span> min_height</span>
<span id="cb1-246"><a href="#cb1-246" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.is_forward <span class="op">=</span> is_forward</span>
<span id="cb1-247"><a href="#cb1-247" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-248"><a href="#cb1-248" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-249"><a href="#cb1-249" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> color(<span class="va">self</span>):</span>
<span id="cb1-250"><a href="#cb1-250" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Return color based on marker's direction and target base</span></span>
<span id="cb1-251"><a href="#cb1-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-252"><a href="#cb1-252" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-253"><a href="#cb1-253" aria-hidden="true" tabindex="-1"></a><span class="co">            str: color</span></span>
<span id="cb1-254"><a href="#cb1-254" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-255"><a href="#cb1-255" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> forward_color.get(<span class="va">self</span>.base) <span class="cf">if</span> <span class="va">self</span>.is_forward <span class="cf">else</span> reverse_color.get(<span class="va">self</span>.base)</span>
<span id="cb1-256"><a href="#cb1-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-257"><a href="#cb1-257" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-258"><a href="#cb1-258" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> data_chanel(<span class="va">self</span>):</span>
<span id="cb1-259"><a href="#cb1-259" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Return data chanel based on marker's direction and target base</span></span>
<span id="cb1-260"><a href="#cb1-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-261"><a href="#cb1-261" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-262"><a href="#cb1-262" aria-hidden="true" tabindex="-1"></a><span class="co">            _type_: _description_</span></span>
<span id="cb1-263"><a href="#cb1-263" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-264"><a href="#cb1-264" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> data_chanel_map.get(<span class="va">self</span>.base) <span class="cf">if</span> <span class="va">self</span>.is_forward <span class="cf">else</span> data_chanel_map.get(reverse_map.get(<span class="va">self</span>.base))</span>
<span id="cb1-265"><a href="#cb1-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-266"><a href="#cb1-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-267"><a href="#cb1-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-268"><a href="#cb1-268" aria-hidden="true" tabindex="-1"></a><span class="co"># define the color of forward</span></span>
<span id="cb1-269"><a href="#cb1-269" aria-hidden="true" tabindex="-1"></a>forward_color <span class="op">=</span> {</span>
<span id="cb1-270"><a href="#cb1-270" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Ref'</span>: <span class="st">'orange'</span>,</span>
<span id="cb1-271"><a href="#cb1-271" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A'</span>: <span class="st">'green'</span>,</span>
<span id="cb1-272"><a href="#cb1-272" aria-hidden="true" tabindex="-1"></a>    <span class="st">'T'</span>: <span class="st">'red'</span>,</span>
<span id="cb1-273"><a href="#cb1-273" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C'</span>: <span class="st">'black'</span>,</span>
<span id="cb1-274"><a href="#cb1-274" aria-hidden="true" tabindex="-1"></a>    <span class="st">'G'</span>: <span class="st">'blue'</span></span>
<span id="cb1-275"><a href="#cb1-275" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-276"><a href="#cb1-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-277"><a href="#cb1-277" aria-hidden="true" tabindex="-1"></a><span class="co"># define the color of reverse</span></span>
<span id="cb1-278"><a href="#cb1-278" aria-hidden="true" tabindex="-1"></a>reverse_color <span class="op">=</span> {</span>
<span id="cb1-279"><a href="#cb1-279" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Ref'</span>: <span class="st">'orange'</span>,</span>
<span id="cb1-280"><a href="#cb1-280" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A'</span>: <span class="st">'red'</span>,</span>
<span id="cb1-281"><a href="#cb1-281" aria-hidden="true" tabindex="-1"></a>    <span class="st">'T'</span>: <span class="st">'green'</span>,</span>
<span id="cb1-282"><a href="#cb1-282" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C'</span>: <span class="st">'blue'</span>,</span>
<span id="cb1-283"><a href="#cb1-283" aria-hidden="true" tabindex="-1"></a>    <span class="st">'G'</span>: <span class="st">'black'</span></span>
<span id="cb1-284"><a href="#cb1-284" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-285"><a href="#cb1-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-286"><a href="#cb1-286" aria-hidden="true" tabindex="-1"></a><span class="co"># define data chanel</span></span>
<span id="cb1-287"><a href="#cb1-287" aria-hidden="true" tabindex="-1"></a>data_chanel_map <span class="op">=</span> {</span>
<span id="cb1-288"><a href="#cb1-288" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Ref'</span>: <span class="st">'DATA105'</span>,</span>
<span id="cb1-289"><a href="#cb1-289" aria-hidden="true" tabindex="-1"></a>    <span class="st">'G'</span>:<span class="st">'DATA1'</span>,</span>
<span id="cb1-290"><a href="#cb1-290" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A'</span>:<span class="st">'DATA2'</span>,</span>
<span id="cb1-291"><a href="#cb1-291" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C'</span>:<span class="st">'DATA3'</span>,</span>
<span id="cb1-292"><a href="#cb1-292" aria-hidden="true" tabindex="-1"></a>    <span class="st">'T'</span>:<span class="st">'DATA4'</span></span>
<span id="cb1-293"><a href="#cb1-293" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-294"><a href="#cb1-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-295"><a href="#cb1-295" aria-hidden="true" tabindex="-1"></a><span class="co"># revese complementation of base.</span></span>
<span id="cb1-296"><a href="#cb1-296" aria-hidden="true" tabindex="-1"></a>reverse_map <span class="op">=</span> {</span>
<span id="cb1-297"><a href="#cb1-297" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A'</span>:<span class="st">'T'</span>,</span>
<span id="cb1-298"><a href="#cb1-298" aria-hidden="true" tabindex="-1"></a>    <span class="st">'T'</span>:<span class="st">'A'</span>,</span>
<span id="cb1-299"><a href="#cb1-299" aria-hidden="true" tabindex="-1"></a>    <span class="st">'G'</span>:<span class="st">'C'</span>,</span>
<span id="cb1-300"><a href="#cb1-300" aria-hidden="true" tabindex="-1"></a>    <span class="st">'C'</span>:<span class="st">'G'</span></span>
<span id="cb1-301"><a href="#cb1-301" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-302"><a href="#cb1-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-303"><a href="#cb1-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-304"><a href="#cb1-304" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Marker:</span>
<span id="cb1-305"><a href="#cb1-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-306"><a href="#cb1-306" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, gene:<span class="bu">str</span>, marker_name: <span class="bu">str</span>, marker_label:<span class="bu">str</span>, panel:<span class="bu">str</span>, is_forward:<span class="bu">bool</span>, alleles <span class="op">=</span> []) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-307"><a href="#cb1-307" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Creat a marker class</span></span>
<span id="cb1-308"><a href="#cb1-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-309"><a href="#cb1-309" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-310"><a href="#cb1-310" aria-hidden="true" tabindex="-1"></a><span class="co">            gene (str): gene name</span></span>
<span id="cb1-311"><a href="#cb1-311" aria-hidden="true" tabindex="-1"></a><span class="co">            marker_name (str): marker name</span></span>
<span id="cb1-312"><a href="#cb1-312" aria-hidden="true" tabindex="-1"></a><span class="co">            marker_label (str): marker label</span></span>
<span id="cb1-313"><a href="#cb1-313" aria-hidden="true" tabindex="-1"></a><span class="co">            panel (str): panel name, it should be identical with chanel detected from FSA file</span></span>
<span id="cb1-314"><a href="#cb1-314" aria-hidden="true" tabindex="-1"></a><span class="co">            is_forward (bool): direction of this marker, True mean forward, False mean reverse.</span></span>
<span id="cb1-315"><a href="#cb1-315" aria-hidden="true" tabindex="-1"></a><span class="co">            alleles (list, optional): Allele class information. Defaults to [].</span></span>
<span id="cb1-316"><a href="#cb1-316" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-317"><a href="#cb1-317" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.gene <span class="op">=</span> gene</span>
<span id="cb1-318"><a href="#cb1-318" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.marker_name <span class="op">=</span> marker_name</span>
<span id="cb1-319"><a href="#cb1-319" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.marker_label <span class="op">=</span> marker_label</span>
<span id="cb1-320"><a href="#cb1-320" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.is_forward <span class="op">=</span> is_forward</span>
<span id="cb1-321"><a href="#cb1-321" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.alleles <span class="op">=</span> alleles</span>
<span id="cb1-322"><a href="#cb1-322" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.panel <span class="op">=</span> panel</span>
<span id="cb1-323"><a href="#cb1-323" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.direction <span class="op">=</span> <span class="st">'Forward'</span> <span class="cf">if</span> <span class="va">self</span>.is_forward <span class="cf">else</span> <span class="st">'Reverse'</span></span>
<span id="cb1-324"><a href="#cb1-324" aria-hidden="true" tabindex="-1"></a>        <span class="co"># self.num_alleles = 0</span></span>
<span id="cb1-325"><a href="#cb1-325" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-326"><a href="#cb1-326" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> add_allele(<span class="va">self</span>, allele):</span>
<span id="cb1-327"><a href="#cb1-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-328"><a href="#cb1-328" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Update allele to the list of allele in this marker.</span></span>
<span id="cb1-329"><a href="#cb1-329" aria-hidden="true" tabindex="-1"></a><span class="co">            It also update the direction of the marker to this allele.</span></span>
<span id="cb1-330"><a href="#cb1-330" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-331"><a href="#cb1-331" aria-hidden="true" tabindex="-1"></a>        allele.is_forward <span class="op">=</span> <span class="va">self</span>.is_forward</span>
<span id="cb1-332"><a href="#cb1-332" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.alleles.append(allele)</span>
<span id="cb1-333"><a href="#cb1-333" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-334"><a href="#cb1-334" aria-hidden="true" tabindex="-1"></a>    <span class="at">@property</span></span>
<span id="cb1-335"><a href="#cb1-335" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> num_alleles(<span class="va">self</span>):</span>
<span id="cb1-336"><a href="#cb1-336" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="bu">len</span>(<span class="va">self</span>.alleles)</span>
<span id="cb1-337"><a href="#cb1-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-338"><a href="#cb1-338" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> update_allele_config(<span class="va">self</span>, allele_number:<span class="bu">int</span>, min_bin:<span class="bu">int</span><span class="op">=</span><span class="va">None</span>, max_bin:<span class="bu">int</span><span class="op">=</span><span class="va">None</span>, min_height:<span class="bu">int</span><span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-339"><a href="#cb1-339" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""update allele configuration to adjust the bin size and height to call the peaks</span></span>
<span id="cb1-340"><a href="#cb1-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-341"><a href="#cb1-341" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-342"><a href="#cb1-342" aria-hidden="true" tabindex="-1"></a><span class="co">            allele_number (int): number of allele in the marker. It ussually can be 0,1 since each marker often have 2 alleles, but somes may have more than 2.</span></span>
<span id="cb1-343"><a href="#cb1-343" aria-hidden="true" tabindex="-1"></a><span class="co">            min_bin (int, optional): Min bin size for this allele. Defaults to None.</span></span>
<span id="cb1-344"><a href="#cb1-344" aria-hidden="true" tabindex="-1"></a><span class="co">            max_bin (int, optional): Max bin sizx for this allele. Defaults to None.</span></span>
<span id="cb1-345"><a href="#cb1-345" aria-hidden="true" tabindex="-1"></a><span class="co">            min_height (int, optional): Min height for this allele. Defaults to None.</span></span>
<span id="cb1-346"><a href="#cb1-346" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-347"><a href="#cb1-347" aria-hidden="true" tabindex="-1"></a>        allele <span class="op">=</span> <span class="va">self</span>.alleles[allele_number]</span>
<span id="cb1-348"><a href="#cb1-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-349"><a href="#cb1-349" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> min_bin:</span>
<span id="cb1-350"><a href="#cb1-350" aria-hidden="true" tabindex="-1"></a>            allele.min_bin <span class="op">=</span> min_bin</span>
<span id="cb1-351"><a href="#cb1-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-352"><a href="#cb1-352" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> max_bin:</span>
<span id="cb1-353"><a href="#cb1-353" aria-hidden="true" tabindex="-1"></a>            allele.max_bin <span class="op">=</span> max_bin</span>
<span id="cb1-354"><a href="#cb1-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-355"><a href="#cb1-355" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> min_height:</span>
<span id="cb1-356"><a href="#cb1-356" aria-hidden="true" tabindex="-1"></a>            allele.min_height <span class="op">=</span> min_height</span>
<span id="cb1-357"><a href="#cb1-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-358"><a href="#cb1-358" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.alleles[allele_number] <span class="op">=</span> allele</span>
<span id="cb1-359"><a href="#cb1-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-360"><a href="#cb1-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-361"><a href="#cb1-361" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_index_by_base_range(base_pred, base_ranges<span class="op">=</span>(<span class="dv">0</span>,<span class="dv">200</span>)):</span>
<span id="cb1-362"><a href="#cb1-362" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Get index of orginal intensity based on intested basepairs.</span></span>
<span id="cb1-363"><a href="#cb1-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-364"><a href="#cb1-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-365"><a href="#cb1-365" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-366"><a href="#cb1-366" aria-hidden="true" tabindex="-1"></a><span class="co">        base_pred (_type_): _description_</span></span>
<span id="cb1-367"><a href="#cb1-367" aria-hidden="true" tabindex="-1"></a><span class="co">        ranges (tuple, optional): _description_. Defaults to (0,200).</span></span>
<span id="cb1-368"><a href="#cb1-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-369"><a href="#cb1-369" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-370"><a href="#cb1-370" aria-hidden="true" tabindex="-1"></a><span class="co">        _type_: _description_</span></span>
<span id="cb1-371"><a href="#cb1-371" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-372"><a href="#cb1-372" aria-hidden="true" tabindex="-1"></a>    min_index <span class="op">=</span> np.where(base_pred <span class="op">&gt;=</span> base_ranges[<span class="dv">0</span>])[<span class="dv">0</span>][<span class="dv">0</span>]</span>
<span id="cb1-373"><a href="#cb1-373" aria-hidden="true" tabindex="-1"></a>    max_index <span class="op">=</span> np.where(base_pred <span class="op">&gt;=</span> base_ranges[<span class="dv">1</span>])[<span class="dv">0</span>][<span class="dv">0</span>]</span>
<span id="cb1-374"><a href="#cb1-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-375"><a href="#cb1-375" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> (min_index, max_index)</span>
<span id="cb1-376"><a href="#cb1-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-377"><a href="#cb1-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-378"><a href="#cb1-378" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_raw_intensity(fsa,min_index, max_index, select_base, base_range<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">80</span>)):</span>
<span id="cb1-379"><a href="#cb1-379" aria-hidden="true" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize <span class="op">=</span>(<span class="dv">12</span>,<span class="dv">3</span>), dpi <span class="op">=</span> <span class="dv">200</span>)</span>
<span id="cb1-380"><a href="#cb1-380" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="op">*</span>base_range)</span>
<span id="cb1-381"><a href="#cb1-381" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> base <span class="kw">in</span> data_chanel_map:</span>
<span id="cb1-382"><a href="#cb1-382" aria-hidden="true" tabindex="-1"></a>        chanel <span class="op">=</span> data_chanel_map.get(base)</span>
<span id="cb1-383"><a href="#cb1-383" aria-hidden="true" tabindex="-1"></a>        intensity <span class="op">=</span> fsa.extract_intensity(chanel, (min_index, max_index))</span>
<span id="cb1-384"><a href="#cb1-384" aria-hidden="true" tabindex="-1"></a>        color <span class="op">=</span> forward_color.get(base)</span>
<span id="cb1-385"><a href="#cb1-385" aria-hidden="true" tabindex="-1"></a>        alpha <span class="op">=</span> <span class="fl">0.6</span> <span class="cf">if</span> base <span class="op">==</span> <span class="st">'Ref'</span> <span class="cf">else</span> <span class="dv">1</span></span>
<span id="cb1-386"><a href="#cb1-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-387"><a href="#cb1-387" aria-hidden="true" tabindex="-1"></a>        ax.plot(select_base, intensity, color <span class="op">=</span> color, label <span class="op">=</span> chanel, alpha <span class="op">=</span> alpha)</span>
<span id="cb1-388"><a href="#cb1-388" aria-hidden="true" tabindex="-1"></a>    plt.legend(loc<span class="op">=</span><span class="st">'upper left'</span>)</span>
<span id="cb1-389"><a href="#cb1-389" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb1-390"><a href="#cb1-390" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb1-391"><a href="#cb1-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-392"><a href="#cb1-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-393"><a href="#cb1-393" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_markers(peak_table):</span>
<span id="cb1-394"><a href="#cb1-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-395"><a href="#cb1-395" aria-hidden="true" tabindex="-1"></a>    maker_list <span class="op">=</span> {}</span>
<span id="cb1-396"><a href="#cb1-396" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> _,row <span class="kw">in</span> peak_table.iterrows():</span>
<span id="cb1-397"><a href="#cb1-397" aria-hidden="true" tabindex="-1"></a>        m <span class="op">=</span> Marker(row.gene, row.marker, row.marker_label, row.panel, row.is_forward, [])</span>
<span id="cb1-398"><a href="#cb1-398" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-399"><a href="#cb1-399" aria-hidden="true" tabindex="-1"></a>        m.define_colors <span class="op">=</span> {row.w_base:row.w_color, row.m_base:row.m_color}</span>
<span id="cb1-400"><a href="#cb1-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-401"><a href="#cb1-401" aria-hidden="true" tabindex="-1"></a>        a1 <span class="op">=</span> Allele(row.w_base, <span class="st">'wildtype'</span>, row.is_forward, row.w_min, row.w_max)</span>
<span id="cb1-402"><a href="#cb1-402" aria-hidden="true" tabindex="-1"></a>        <span class="co"># a1.set_color(row.is_forward)</span></span>
<span id="cb1-403"><a href="#cb1-403" aria-hidden="true" tabindex="-1"></a>        a2 <span class="op">=</span> Allele(row.m_base, <span class="st">'mutant'</span>, row.is_forward, row.m_min, row.m_max)</span>
<span id="cb1-404"><a href="#cb1-404" aria-hidden="true" tabindex="-1"></a>        <span class="co"># a2.set_color(row.is_forward)</span></span>
<span id="cb1-405"><a href="#cb1-405" aria-hidden="true" tabindex="-1"></a>        m.add_allele(a1)</span>
<span id="cb1-406"><a href="#cb1-406" aria-hidden="true" tabindex="-1"></a>        m.add_allele(a2)</span>
<span id="cb1-407"><a href="#cb1-407" aria-hidden="true" tabindex="-1"></a>        maker_list[m.marker_name] <span class="op">=</span> m</span>
<span id="cb1-408"><a href="#cb1-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-409"><a href="#cb1-409" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> maker_list</span>
<span id="cb1-410"><a href="#cb1-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-411"><a href="#cb1-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-412"><a href="#cb1-412" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> finding_peak_by_markers(markers:<span class="bu">dict</span>, fsa:FSA, panel:<span class="bu">str</span>, select_base:<span class="bu">list</span>, index_range:<span class="bu">tuple</span>):</span>
<span id="cb1-413"><a href="#cb1-413" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Finding peak by marker dictionary and update the called result to the marker dictionary</span></span>
<span id="cb1-414"><a href="#cb1-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-415"><a href="#cb1-415" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-416"><a href="#cb1-416" aria-hidden="true" tabindex="-1"></a><span class="co">        markers (dict): A dictionary of markers that includes defined information to call a peak such as bin size, height, data chanel and base </span></span>
<span id="cb1-417"><a href="#cb1-417" aria-hidden="true" tabindex="-1"></a><span class="co">        fsa (FSA): FSA data loaded from fsa file</span></span>
<span id="cb1-418"><a href="#cb1-418" aria-hidden="true" tabindex="-1"></a><span class="co">        panel (str): Panel definition, it should be match with panel defined in the file name, need to be discussed with PGx to make consensus naming convention.</span></span>
<span id="cb1-419"><a href="#cb1-419" aria-hidden="true" tabindex="-1"></a><span class="co">        select_base (list): a list of predicted basepair from model, it was used to match with bin size and data points.</span></span>
<span id="cb1-420"><a href="#cb1-420" aria-hidden="true" tabindex="-1"></a><span class="co">        index_range (tuple): index data point to extract intensity data</span></span>
<span id="cb1-421"><a href="#cb1-421" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-422"><a href="#cb1-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-423"><a href="#cb1-423" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> marker_name <span class="kw">in</span> markers:</span>
<span id="cb1-424"><a href="#cb1-424" aria-hidden="true" tabindex="-1"></a>        marker <span class="op">=</span> markers.get(marker_name)</span>
<span id="cb1-425"><a href="#cb1-425" aria-hidden="true" tabindex="-1"></a>        marker.sample_name <span class="op">=</span> fsa.sample_name</span>
<span id="cb1-426"><a href="#cb1-426" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> marker.panel <span class="op">==</span> panel:</span>
<span id="cb1-427"><a href="#cb1-427" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i,allele <span class="kw">in</span> <span class="bu">enumerate</span>(marker.alleles):</span>
<span id="cb1-428"><a href="#cb1-428" aria-hidden="true" tabindex="-1"></a>                intensity <span class="op">=</span> fsa.extract_intensity(allele.data_chanel, (index_range[<span class="dv">0</span>],index_range[<span class="dv">1</span>]))</span>
<span id="cb1-429"><a href="#cb1-429" aria-hidden="true" tabindex="-1"></a>                _min <span class="op">=</span> np.where(select_base <span class="op">&gt;=</span> allele.min_bin)[<span class="dv">0</span>][<span class="dv">0</span>]</span>
<span id="cb1-430"><a href="#cb1-430" aria-hidden="true" tabindex="-1"></a>                _max <span class="op">=</span> np.where(select_base <span class="op">&lt;=</span> allele.max_bin)[<span class="dv">0</span>][<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb1-431"><a href="#cb1-431" aria-hidden="true" tabindex="-1"></a>                i_base <span class="op">=</span> select_base[_min:_max]</span>
<span id="cb1-432"><a href="#cb1-432" aria-hidden="true" tabindex="-1"></a>                i_intensity <span class="op">=</span> intensity[_min:_max]</span>
<span id="cb1-433"><a href="#cb1-433" aria-hidden="true" tabindex="-1"></a>                peak,height <span class="op">=</span> finding_peaks(i_intensity, min_height<span class="op">=</span> allele.min_height, min_width<span class="op">=</span><span class="fl">0.7</span>, min_prominence<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb1-434"><a href="#cb1-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-435"><a href="#cb1-435" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> peak.size <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb1-436"><a href="#cb1-436" aria-hidden="true" tabindex="-1"></a>                    allele.is_detected <span class="op">=</span> <span class="va">True</span></span>
<span id="cb1-437"><a href="#cb1-437" aria-hidden="true" tabindex="-1"></a>                    allele.peak <span class="op">=</span> peak</span>
<span id="cb1-438"><a href="#cb1-438" aria-hidden="true" tabindex="-1"></a>                    allele.size <span class="op">=</span> i_base[peak]</span>
<span id="cb1-439"><a href="#cb1-439" aria-hidden="true" tabindex="-1"></a>                    allele.height <span class="op">=</span> height</span>
<span id="cb1-440"><a href="#cb1-440" aria-hidden="true" tabindex="-1"></a>                    allele.status <span class="op">=</span> <span class="st">'ok'</span></span>
<span id="cb1-441"><a href="#cb1-441" aria-hidden="true" tabindex="-1"></a>                    allele.message <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-442"><a href="#cb1-442" aria-hidden="true" tabindex="-1"></a>                <span class="cf">elif</span> peak.size <span class="op">&gt;</span> <span class="dv">1</span>:</span>
<span id="cb1-443"><a href="#cb1-443" aria-hidden="true" tabindex="-1"></a>                    allele.is_detected <span class="op">=</span> <span class="va">True</span></span>
<span id="cb1-444"><a href="#cb1-444" aria-hidden="true" tabindex="-1"></a>                    allele.peak <span class="op">=</span> peak</span>
<span id="cb1-445"><a href="#cb1-445" aria-hidden="true" tabindex="-1"></a>                    allele.size <span class="op">=</span> i_base[peak]</span>
<span id="cb1-446"><a href="#cb1-446" aria-hidden="true" tabindex="-1"></a>                    allele.height <span class="op">=</span> height</span>
<span id="cb1-447"><a href="#cb1-447" aria-hidden="true" tabindex="-1"></a>                    allele.status <span class="op">=</span> <span class="st">'warning'</span></span>
<span id="cb1-448"><a href="#cb1-448" aria-hidden="true" tabindex="-1"></a>                    allele.message <span class="op">=</span> <span class="st">'More than 1 peaks detected!'</span></span>
<span id="cb1-449"><a href="#cb1-449" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb1-450"><a href="#cb1-450" aria-hidden="true" tabindex="-1"></a>                    allele.is_detected <span class="op">=</span> <span class="va">False</span></span>
<span id="cb1-451"><a href="#cb1-451" aria-hidden="true" tabindex="-1"></a>                    allele.peak <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-452"><a href="#cb1-452" aria-hidden="true" tabindex="-1"></a>                    allele.size <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-453"><a href="#cb1-453" aria-hidden="true" tabindex="-1"></a>                    allele.height <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-454"><a href="#cb1-454" aria-hidden="true" tabindex="-1"></a>                    allele.status <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-455"><a href="#cb1-455" aria-hidden="true" tabindex="-1"></a>                    allele.message <span class="op">=</span> <span class="st">'Peak(s) could not be detected. Please check peak ranges if required!'</span></span>
<span id="cb1-456"><a href="#cb1-456" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb1-457"><a href="#cb1-457" aria-hidden="true" tabindex="-1"></a>                <span class="co"># print('append allele', marker.marker_name)</span></span>
<span id="cb1-458"><a href="#cb1-458" aria-hidden="true" tabindex="-1"></a>                marker.alleles[i] <span class="op">=</span> allele</span>
<span id="cb1-459"><a href="#cb1-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-460"><a href="#cb1-460" aria-hidden="true" tabindex="-1"></a>            markers[marker_name] <span class="op">=</span> marker</span>
<span id="cb1-461"><a href="#cb1-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-462"><a href="#cb1-462" aria-hidden="true" tabindex="-1"></a>    <span class="co"># update genotype </span></span>
<span id="cb1-463"><a href="#cb1-463" aria-hidden="true" tabindex="-1"></a>    update_marker_genotype(markers, panel)</span>
<span id="cb1-464"><a href="#cb1-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-465"><a href="#cb1-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-466"><a href="#cb1-466" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> update_marker_genotype(markers:<span class="bu">dict</span>, panel:<span class="bu">str</span>):</span>
<span id="cb1-467"><a href="#cb1-467" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Update the genotype of each marker based on the peak calling process.</span></span>
<span id="cb1-468"><a href="#cb1-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-469"><a href="#cb1-469" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-470"><a href="#cb1-470" aria-hidden="true" tabindex="-1"></a><span class="co">        markers (dict): a dictionary of a called marker</span></span>
<span id="cb1-471"><a href="#cb1-471" aria-hidden="true" tabindex="-1"></a><span class="co">        panel (str): panel</span></span>
<span id="cb1-472"><a href="#cb1-472" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-473"><a href="#cb1-473" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> marker_name <span class="kw">in</span> markers:</span>
<span id="cb1-474"><a href="#cb1-474" aria-hidden="true" tabindex="-1"></a>        marker <span class="op">=</span> markers.get(marker_name)</span>
<span id="cb1-475"><a href="#cb1-475" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> marker.panel <span class="op">==</span> panel:</span>
<span id="cb1-476"><a href="#cb1-476" aria-hidden="true" tabindex="-1"></a>            genotype <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-477"><a href="#cb1-477" aria-hidden="true" tabindex="-1"></a>            phenotype <span class="op">=</span> <span class="st">''</span></span>
<span id="cb1-478"><a href="#cb1-478" aria-hidden="true" tabindex="-1"></a>            is_called <span class="op">=</span> <span class="va">False</span></span>
<span id="cb1-479"><a href="#cb1-479" aria-hidden="true" tabindex="-1"></a>            basetype <span class="op">=</span> []</span>
<span id="cb1-480"><a href="#cb1-480" aria-hidden="true" tabindex="-1"></a>            called_base <span class="op">=</span> []</span>
<span id="cb1-481"><a href="#cb1-481" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> j, allele <span class="kw">in</span> <span class="bu">enumerate</span>(marker.alleles):</span>
<span id="cb1-482"><a href="#cb1-482" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> allele.status <span class="op">==</span> <span class="st">'ok'</span>:</span>
<span id="cb1-483"><a href="#cb1-483" aria-hidden="true" tabindex="-1"></a>                    called_base.append(allele.base)</span>
<span id="cb1-484"><a href="#cb1-484" aria-hidden="true" tabindex="-1"></a>                    basetype.append(allele.basetype)</span>
<span id="cb1-485"><a href="#cb1-485" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb1-486"><a href="#cb1-486" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(called_base) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb1-487"><a href="#cb1-487" aria-hidden="true" tabindex="-1"></a>                genotype <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>called_base[<span class="dv">0</span>]<span class="sc">}{</span>called_base[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb1-488"><a href="#cb1-488" aria-hidden="true" tabindex="-1"></a>                is_called <span class="op">=</span> <span class="va">True</span></span>
<span id="cb1-489"><a href="#cb1-489" aria-hidden="true" tabindex="-1"></a>            <span class="cf">elif</span> <span class="bu">len</span>(called_base) <span class="op">==</span> <span class="dv">2</span>:</span>
<span id="cb1-490"><a href="#cb1-490" aria-hidden="true" tabindex="-1"></a>                genotype <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>called_base[<span class="dv">0</span>]<span class="sc">}{</span>called_base[<span class="dv">1</span>]<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb1-491"><a href="#cb1-491" aria-hidden="true" tabindex="-1"></a>                is_called <span class="op">=</span> <span class="va">True</span></span>
<span id="cb1-492"><a href="#cb1-492" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb1-493"><a href="#cb1-493" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(<span class="bu">set</span>(basetype)) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb1-494"><a href="#cb1-494" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> basetype[<span class="dv">0</span>] <span class="op">==</span> <span class="st">'wildtype'</span>:</span>
<span id="cb1-495"><a href="#cb1-495" aria-hidden="true" tabindex="-1"></a>                    phenotype <span class="op">=</span> <span class="st">'wildtype'</span></span>
<span id="cb1-496"><a href="#cb1-496" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb1-497"><a href="#cb1-497" aria-hidden="true" tabindex="-1"></a>                    phenotype <span class="op">=</span> <span class="st">'homozygous mutant'</span></span>
<span id="cb1-498"><a href="#cb1-498" aria-hidden="true" tabindex="-1"></a>            <span class="cf">elif</span> <span class="bu">len</span>(<span class="bu">set</span>(basetype)) <span class="op">==</span> <span class="dv">2</span> <span class="kw">and</span> <span class="bu">len</span>(basetype) <span class="op">==</span> <span class="dv">2</span>:</span>
<span id="cb1-499"><a href="#cb1-499" aria-hidden="true" tabindex="-1"></a>                phenotype <span class="op">=</span> <span class="st">'heterozygous'</span></span>
<span id="cb1-500"><a href="#cb1-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-501"><a href="#cb1-501" aria-hidden="true" tabindex="-1"></a>            marker.genotype <span class="op">=</span> genotype</span>
<span id="cb1-502"><a href="#cb1-502" aria-hidden="true" tabindex="-1"></a>            marker.is_called <span class="op">=</span> is_called</span>
<span id="cb1-503"><a href="#cb1-503" aria-hidden="true" tabindex="-1"></a>            marker.phenotype <span class="op">=</span> phenotype</span>
<span id="cb1-504"><a href="#cb1-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-505"><a href="#cb1-505" aria-hidden="true" tabindex="-1"></a>            markers[marker_name] <span class="op">=</span> marker</span>
<span id="cb1-506"><a href="#cb1-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-507"><a href="#cb1-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-508"><a href="#cb1-508" aria-hidden="true" tabindex="-1"></a><span class="co"># def print_marker(markers):</span></span>
<span id="cb1-509"><a href="#cb1-509" aria-hidden="true" tabindex="-1"></a><span class="co">#     status = False</span></span>
<span id="cb1-510"><a href="#cb1-510" aria-hidden="true" tabindex="-1"></a><span class="co">#     for marker_name in markers:</span></span>
<span id="cb1-511"><a href="#cb1-511" aria-hidden="true" tabindex="-1"></a><span class="co">#         marker = markers.get(marker_name)</span></span>
<span id="cb1-512"><a href="#cb1-512" aria-hidden="true" tabindex="-1"></a><span class="co">#         if hasattr(marker, 'is_called'):</span></span>
<span id="cb1-513"><a href="#cb1-513" aria-hidden="true" tabindex="-1"></a><span class="co">#             status = marker.is_called</span></span>
<span id="cb1-514"><a href="#cb1-514" aria-hidden="true" tabindex="-1"></a><span class="co">#             print(marker.marker_name, marker.panel, status, marker.genotype, marker.phenotype)</span></span>
<span id="cb1-515"><a href="#cb1-515" aria-hidden="true" tabindex="-1"></a><span class="co">#         else:</span></span>
<span id="cb1-516"><a href="#cb1-516" aria-hidden="true" tabindex="-1"></a><span class="co">#             print(marker.marker_name, marker.panel)</span></span>
<span id="cb1-517"><a href="#cb1-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-518"><a href="#cb1-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-519"><a href="#cb1-519" aria-hidden="true" tabindex="-1"></a><span class="co"># def load_snapshot_file(snapshot_file):</span></span>
<span id="cb1-520"><a href="#cb1-520" aria-hidden="true" tabindex="-1"></a><span class="co">#     df = pd.read_csv(snapshot_file, sep='\t').fillna("")</span></span>
<span id="cb1-521"><a href="#cb1-521" aria-hidden="true" tabindex="-1"></a><span class="co">#     df['Allele 2'] = np.where(df['Allele 2'] == "", df['Allele 1'], df['Allele 2'])</span></span>
<span id="cb1-522"><a href="#cb1-522" aria-hidden="true" tabindex="-1"></a><span class="co">#     df["Genotype"] = df['Allele 1'] + df["Allele 2"]</span></span>
<span id="cb1-523"><a href="#cb1-523" aria-hidden="true" tabindex="-1"></a><span class="co">#     df['Gene'] = [x[0] for x in df['Marker'].str.split("_")]</span></span>
<span id="cb1-524"><a href="#cb1-524" aria-hidden="true" tabindex="-1"></a><span class="co">#     df['ExtractedSample'] = [extract_sample_name(x) for x in df['Sample Name']]</span></span>
<span id="cb1-525"><a href="#cb1-525" aria-hidden="true" tabindex="-1"></a><span class="co">#     df['SampleStatus'] = [is_valid_sample_name(x) for x in df['Sample Name']]</span></span>
<span id="cb1-526"><a href="#cb1-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-527"><a href="#cb1-527" aria-hidden="true" tabindex="-1"></a><span class="co">#     df = df[['Sample File', 'Sample Name', 'ExtractedSample', 'SampleStatus', 'Panel',</span></span>
<span id="cb1-528"><a href="#cb1-528" aria-hidden="true" tabindex="-1"></a><span class="co">#              'Gene', 'Marker', 'Genotype', 'Allele 1', 'Allele 2']]</span></span>
<span id="cb1-529"><a href="#cb1-529" aria-hidden="true" tabindex="-1"></a><span class="co">#     return df</span></span>
<span id="cb1-530"><a href="#cb1-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-531"><a href="#cb1-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-532"><a href="#cb1-532" aria-hidden="true" tabindex="-1"></a><span class="co"># def is_valid_sample_name(sample_name, p='_S[0-9]+'):</span></span>
<span id="cb1-533"><a href="#cb1-533" aria-hidden="true" tabindex="-1"></a><span class="co">#     pattern = re.compile(p)</span></span>
<span id="cb1-534"><a href="#cb1-534" aria-hidden="true" tabindex="-1"></a><span class="co">#     if pattern.search(sample_name):</span></span>
<span id="cb1-535"><a href="#cb1-535" aria-hidden="true" tabindex="-1"></a><span class="co">#         return True</span></span>
<span id="cb1-536"><a href="#cb1-536" aria-hidden="true" tabindex="-1"></a><span class="co">#     else:</span></span>
<span id="cb1-537"><a href="#cb1-537" aria-hidden="true" tabindex="-1"></a><span class="co">#         return False</span></span>
<span id="cb1-538"><a href="#cb1-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-539"><a href="#cb1-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-540"><a href="#cb1-540" aria-hidden="true" tabindex="-1"></a><span class="co"># def format_call_markers(markers):</span></span>
<span id="cb1-541"><a href="#cb1-541" aria-hidden="true" tabindex="-1"></a><span class="co">#     df = pd.DataFrame()</span></span>
<span id="cb1-542"><a href="#cb1-542" aria-hidden="true" tabindex="-1"></a><span class="co">#     for marker_name in markers:</span></span>
<span id="cb1-543"><a href="#cb1-543" aria-hidden="true" tabindex="-1"></a><span class="co">#         marker = markers.get(marker_name)</span></span>
<span id="cb1-544"><a href="#cb1-544" aria-hidden="true" tabindex="-1"></a><span class="co">#         for allele in marker.alleles:</span></span>
<span id="cb1-545"><a href="#cb1-545" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp = pd.DataFrame(allele.__dict__, index=[0])</span></span>
<span id="cb1-546"><a href="#cb1-546" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['color'] = allele.color</span></span>
<span id="cb1-547"><a href="#cb1-547" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['sample'] = marker.sample_name</span></span>
<span id="cb1-548"><a href="#cb1-548" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['gene'] = marker.gene</span></span>
<span id="cb1-549"><a href="#cb1-549" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['marker'] = marker.marker_name</span></span>
<span id="cb1-550"><a href="#cb1-550" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['panel'] = marker.panel</span></span>
<span id="cb1-551"><a href="#cb1-551" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['direction'] = 'Forward' if marker.is_forward else 'Reverse'</span></span>
<span id="cb1-552"><a href="#cb1-552" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['genotype'] = marker.genotype</span></span>
<span id="cb1-553"><a href="#cb1-553" aria-hidden="true" tabindex="-1"></a><span class="co">#             tmp['phenotype'] = marker.phenotype</span></span>
<span id="cb1-554"><a href="#cb1-554" aria-hidden="true" tabindex="-1"></a><span class="co">#             df = pd.concat([df, tmp])</span></span>
<span id="cb1-555"><a href="#cb1-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-556"><a href="#cb1-556" aria-hidden="true" tabindex="-1"></a><span class="co">#     ## marker table</span></span>
<span id="cb1-557"><a href="#cb1-557" aria-hidden="true" tabindex="-1"></a><span class="co">#     marker_table = (df[df.columns[-7:]].drop_duplicates()).reset_index(drop=True)</span></span>
<span id="cb1-558"><a href="#cb1-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-559"><a href="#cb1-559" aria-hidden="true" tabindex="-1"></a><span class="co">#     # allele table</span></span>
<span id="cb1-560"><a href="#cb1-560" aria-hidden="true" tabindex="-1"></a><span class="co">#     cols = df.columns[:-2]</span></span>
<span id="cb1-561"><a href="#cb1-561" aria-hidden="true" tabindex="-1"></a><span class="co">#     cols = cols[-5:].tolist() + cols[:-5].tolist() </span></span>
<span id="cb1-562"><a href="#cb1-562" aria-hidden="true" tabindex="-1"></a><span class="co">#     allele_table = df[cols].reset_index(drop=True)</span></span>
<span id="cb1-563"><a href="#cb1-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-564"><a href="#cb1-564" aria-hidden="true" tabindex="-1"></a><span class="co">#     return marker_table, allele_table</span></span>
<span id="cb1-565"><a href="#cb1-565" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-566"><a href="#cb1-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-567"><a href="#cb1-567" aria-hidden="true" tabindex="-1"></a><span class="co"># def extract_sample_name(sample_name, p='_S[0-9]+'):</span></span>
<span id="cb1-568"><a href="#cb1-568" aria-hidden="true" tabindex="-1"></a><span class="co">#     pattern = re.compile(p)</span></span>
<span id="cb1-569"><a href="#cb1-569" aria-hidden="true" tabindex="-1"></a><span class="co">#     if pattern.search(sample_name):</span></span>
<span id="cb1-570"><a href="#cb1-570" aria-hidden="true" tabindex="-1"></a><span class="co">#         return pattern.split(sample_name)[0]</span></span>
<span id="cb1-571"><a href="#cb1-571" aria-hidden="true" tabindex="-1"></a><span class="co">#     else:</span></span>
<span id="cb1-572"><a href="#cb1-572" aria-hidden="true" tabindex="-1"></a><span class="co">#         # print(f'{sample_name} is not valid! Keep as origin')</span></span>
<span id="cb1-573"><a href="#cb1-573" aria-hidden="true" tabindex="-1"></a><span class="co">#         return sample_name</span></span>
<span id="cb1-574"><a href="#cb1-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-575"><a href="#cb1-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-576"><a href="#cb1-576" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extract_sample_name(fsa_name:<span class="bu">str</span>, pattern <span class="op">=</span> <span class="st">'[0-9]+-[0-9]+_'</span>):</span>
<span id="cb1-577"><a href="#cb1-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-578"><a href="#cb1-578" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Extract sample name from fsa file name. It can be applied for bin text file as well.</span></span>
<span id="cb1-579"><a href="#cb1-579" aria-hidden="true" tabindex="-1"></a><span class="co">       But we need to define the file name convention with PGx team to have rule to extract the sample name.</span></span>
<span id="cb1-580"><a href="#cb1-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-581"><a href="#cb1-581" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-582"><a href="#cb1-582" aria-hidden="true" tabindex="-1"></a><span class="co">        str: sample name extracted from file name</span></span>
<span id="cb1-583"><a href="#cb1-583" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-584"><a href="#cb1-584" aria-hidden="true" tabindex="-1"></a>    p <span class="op">=</span> re.<span class="bu">compile</span>(pattern)</span>
<span id="cb1-585"><a href="#cb1-585" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> p.search(fsa_name)</span>
<span id="cb1-586"><a href="#cb1-586" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-587"><a href="#cb1-587" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> s.group(<span class="dv">0</span>).replace(<span class="st">'_'</span>, <span class="st">''</span>)</span>
<span id="cb1-588"><a href="#cb1-588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-589"><a href="#cb1-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-590"><a href="#cb1-590" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> GenotypeResult:</span>
<span id="cb1-591"><a href="#cb1-591" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-592"><a href="#cb1-592" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-593"><a href="#cb1-593" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.is_from_file <span class="op">=</span> <span class="va">False</span></span>
<span id="cb1-594"><a href="#cb1-594" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-595"><a href="#cb1-595" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> from_bin_file(<span class="va">self</span>, bin_file:<span class="bu">str</span>, marker_info:<span class="bu">dict</span>):</span>
<span id="cb1-596"><a href="#cb1-596" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Since our result can be obtained from calling FSA file or from bin text file. This function is to process data from bin text file</span></span>
<span id="cb1-597"><a href="#cb1-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-598"><a href="#cb1-598" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-599"><a href="#cb1-599" aria-hidden="true" tabindex="-1"></a><span class="co">            bin_file (str): bin text file obtained from gene mappeer</span></span>
<span id="cb1-600"><a href="#cb1-600" aria-hidden="true" tabindex="-1"></a><span class="co">            marker_info (dict): Information of the defined marker includes color, direction, label and other infomration.</span></span>
<span id="cb1-601"><a href="#cb1-601" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-602"><a href="#cb1-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-603"><a href="#cb1-603" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.<span class="bu">file</span> <span class="op">=</span> bin_file</span>
<span id="cb1-604"><a href="#cb1-604" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.raw_data <span class="op">=</span> pd.read_csv(bin_file, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>).fillna(<span class="st">''</span>)</span>
<span id="cb1-605"><a href="#cb1-605" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.is_from_file <span class="op">=</span> <span class="va">True</span></span>
<span id="cb1-606"><a href="#cb1-606" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.clean_data <span class="op">=</span> <span class="va">self</span>._clean_from_bin_data(marker_info)</span>
<span id="cb1-607"><a href="#cb1-607" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.sample_list <span class="op">=</span> <span class="va">self</span>.clean_data[<span class="st">'sample'</span>].unique().tolist()</span>
<span id="cb1-608"><a href="#cb1-608" aria-hidden="true" tabindex="-1"></a>        <span class="co"># self.marker_table = self._marker_table</span></span>
<span id="cb1-609"><a href="#cb1-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-610"><a href="#cb1-610" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> from_fsa_call(<span class="va">self</span>, called_markers:<span class="bu">dict</span>):</span>
<span id="cb1-611"><a href="#cb1-611" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Process result called from FSA file</span></span>
<span id="cb1-612"><a href="#cb1-612" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-613"><a href="#cb1-613" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-614"><a href="#cb1-614" aria-hidden="true" tabindex="-1"></a><span class="co">            called_markers (dict): marker includes the peaking calling information</span></span>
<span id="cb1-615"><a href="#cb1-615" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-616"><a href="#cb1-616" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.is_from_file <span class="op">=</span> <span class="va">False</span></span>
<span id="cb1-617"><a href="#cb1-617" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.raw_data <span class="op">=</span> called_markers</span>
<span id="cb1-618"><a href="#cb1-618" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-619"><a href="#cb1-619" aria-hidden="true" tabindex="-1"></a>        clean_data <span class="op">=</span> pd.DataFrame()</span>
<span id="cb1-620"><a href="#cb1-620" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> marker_name <span class="kw">in</span> called_markers:</span>
<span id="cb1-621"><a href="#cb1-621" aria-hidden="true" tabindex="-1"></a>            marker <span class="op">=</span> called_markers.get(marker_name)</span>
<span id="cb1-622"><a href="#cb1-622" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> allele <span class="kw">in</span> marker.alleles:</span>
<span id="cb1-623"><a href="#cb1-623" aria-hidden="true" tabindex="-1"></a>                tmp <span class="op">=</span> pd.DataFrame(allele.__dict__, index<span class="op">=</span>[<span class="dv">0</span>])</span>
<span id="cb1-624"><a href="#cb1-624" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'color'</span>] <span class="op">=</span> allele.color</span>
<span id="cb1-625"><a href="#cb1-625" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'sample'</span>] <span class="op">=</span> marker.sample_name</span>
<span id="cb1-626"><a href="#cb1-626" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'gene'</span>] <span class="op">=</span> marker.gene</span>
<span id="cb1-627"><a href="#cb1-627" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'marker'</span>] <span class="op">=</span> marker.marker_name</span>
<span id="cb1-628"><a href="#cb1-628" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'label'</span>] <span class="op">=</span> marker.marker_label</span>
<span id="cb1-629"><a href="#cb1-629" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'panel'</span>] <span class="op">=</span> marker.panel</span>
<span id="cb1-630"><a href="#cb1-630" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'direction'</span>] <span class="op">=</span> <span class="st">'Forward'</span> <span class="cf">if</span> marker.is_forward <span class="cf">else</span> <span class="st">'Reverse'</span></span>
<span id="cb1-631"><a href="#cb1-631" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'genotype'</span>] <span class="op">=</span> marker.genotype</span>
<span id="cb1-632"><a href="#cb1-632" aria-hidden="true" tabindex="-1"></a>                tmp[<span class="st">'phenotype'</span>] <span class="op">=</span> marker.phenotype</span>
<span id="cb1-633"><a href="#cb1-633" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb1-634"><a href="#cb1-634" aria-hidden="true" tabindex="-1"></a>                clean_data <span class="op">=</span> pd.concat([clean_data, tmp])</span>
<span id="cb1-635"><a href="#cb1-635" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-636"><a href="#cb1-636" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.clean_data <span class="op">=</span> clean_data.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb1-637"><a href="#cb1-637" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.sample_list <span class="op">=</span> <span class="va">self</span>.clean_data[<span class="st">'sample'</span>].unique().tolist()</span>
<span id="cb1-638"><a href="#cb1-638" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-639"><a href="#cb1-639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-640"><a href="#cb1-640" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> _clean_from_bin_data(<span class="va">self</span>, marker_info:<span class="bu">dict</span>):</span>
<span id="cb1-641"><a href="#cb1-641" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Detail function to process data from bin text file</span></span>
<span id="cb1-642"><a href="#cb1-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-643"><a href="#cb1-643" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-644"><a href="#cb1-644" aria-hidden="true" tabindex="-1"></a><span class="co">            marker_info (dict): Information of the defined marker includes color, direction, label and other infomration.</span></span>
<span id="cb1-645"><a href="#cb1-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-646"><a href="#cb1-646" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-647"><a href="#cb1-647" aria-hidden="true" tabindex="-1"></a><span class="co">            DataFrame: an updated data frame</span></span>
<span id="cb1-648"><a href="#cb1-648" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-649"><a href="#cb1-649" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-650"><a href="#cb1-650" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="va">self</span>.is_from_file:</span>
<span id="cb1-651"><a href="#cb1-651" aria-hidden="true" tabindex="-1"></a>            cols <span class="op">=</span> [<span class="st">'Sample File'</span>, <span class="st">'Sample Name'</span>, <span class="st">'Marker'</span>, <span class="st">'Panel'</span>, <span class="st">'Allele 1'</span>, <span class="st">'Allele 2'</span>,</span>
<span id="cb1-652"><a href="#cb1-652" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'Size 1'</span>, <span class="st">'Size 2'</span>, <span class="st">'Height 1'</span>, <span class="st">'Height 2'</span>]</span>
<span id="cb1-653"><a href="#cb1-653" aria-hidden="true" tabindex="-1"></a>            m_cols <span class="op">=</span> [<span class="st">'sample_file'</span>, <span class="st">'sample_name'</span>, <span class="st">'marker'</span>, <span class="st">'panel'</span>, <span class="st">'base1'</span>, <span class="st">'base2'</span>,</span>
<span id="cb1-654"><a href="#cb1-654" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'size1'</span>, <span class="st">'size2'</span>, <span class="st">'height1'</span>, <span class="st">'height2'</span>]</span>
<span id="cb1-655"><a href="#cb1-655" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> <span class="va">self</span>.raw_data[cols]</span>
<span id="cb1-656"><a href="#cb1-656" aria-hidden="true" tabindex="-1"></a>            df.columns <span class="op">=</span> m_cols</span>
<span id="cb1-657"><a href="#cb1-657" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb1-658"><a href="#cb1-658" aria-hidden="true" tabindex="-1"></a>            <span class="co"># extract gene, panel and sample</span></span>
<span id="cb1-659"><a href="#cb1-659" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'gene'</span>] <span class="op">=</span> df.marker.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.split(<span class="st">'_'</span>)[<span class="dv">0</span>])</span>
<span id="cb1-660"><a href="#cb1-660" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'panel'</span>] <span class="op">=</span> df.panel.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.split(<span class="st">'-'</span>)[<span class="dv">1</span>])</span>
<span id="cb1-661"><a href="#cb1-661" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'sample'</span>] <span class="op">=</span> df[<span class="st">'sample_name'</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: <span class="st">'-'</span>.join(x.split(<span class="st">'-'</span>)[<span class="dv">1</span>:]))</span>
<span id="cb1-662"><a href="#cb1-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-663"><a href="#cb1-663" aria-hidden="true" tabindex="-1"></a>            <span class="co"># generate genotype</span></span>
<span id="cb1-664"><a href="#cb1-664" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'genotype'</span>] <span class="op">=</span> df.<span class="bu">apply</span>(<span class="kw">lambda</span> row: row[<span class="st">'base1'</span>] <span class="op">+</span> row[<span class="st">'base1'</span>] <span class="cf">if</span> row[<span class="st">'base2'</span>] <span class="op">==</span> <span class="st">''</span> <span class="cf">else</span> row[<span class="st">'base1'</span>] <span class="op">+</span> row[<span class="st">'base2'</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb1-665"><a href="#cb1-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-666"><a href="#cb1-666" aria-hidden="true" tabindex="-1"></a>            <span class="co"># update phenotype</span></span>
<span id="cb1-667"><a href="#cb1-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-668"><a href="#cb1-668" aria-hidden="true" tabindex="-1"></a>            <span class="kw">def</span> _get_type_base(marker):</span>
<span id="cb1-669"><a href="#cb1-669" aria-hidden="true" tabindex="-1"></a>                wildtype_base <span class="op">=</span> []</span>
<span id="cb1-670"><a href="#cb1-670" aria-hidden="true" tabindex="-1"></a>                mutant_base <span class="op">=</span> []</span>
<span id="cb1-671"><a href="#cb1-671" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> allele <span class="kw">in</span> marker.alleles:</span>
<span id="cb1-672"><a href="#cb1-672" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> allele.basetype <span class="op">==</span> <span class="st">'wildtype'</span>:</span>
<span id="cb1-673"><a href="#cb1-673" aria-hidden="true" tabindex="-1"></a>                        wildtype_base.append(allele.base)</span>
<span id="cb1-674"><a href="#cb1-674" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span>:</span>
<span id="cb1-675"><a href="#cb1-675" aria-hidden="true" tabindex="-1"></a>                        mutant_base.append(allele.base)</span>
<span id="cb1-676"><a href="#cb1-676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-677"><a href="#cb1-677" aria-hidden="true" tabindex="-1"></a>                <span class="cf">return</span> wildtype_base, mutant_base</span>
<span id="cb1-678"><a href="#cb1-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-679"><a href="#cb1-679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-680"><a href="#cb1-680" aria-hidden="true" tabindex="-1"></a>            <span class="kw">def</span> _update_phenotype(marker_name, genotype):</span>
<span id="cb1-681"><a href="#cb1-681" aria-hidden="true" tabindex="-1"></a>                marker <span class="op">=</span> marker_info.get(marker_name)</span>
<span id="cb1-682"><a href="#cb1-682" aria-hidden="true" tabindex="-1"></a>                wildtype, mutant <span class="op">=</span> _get_type_base(marker)</span>
<span id="cb1-683"><a href="#cb1-683" aria-hidden="true" tabindex="-1"></a>                _genotype <span class="op">=</span> pd.Series(<span class="bu">list</span>(<span class="bu">set</span>(genotype)))</span>
<span id="cb1-684"><a href="#cb1-684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-685"><a href="#cb1-685" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="bu">len</span>(_genotype) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb1-686"><a href="#cb1-686" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">all</span>(_genotype.isin(wildtype)):</span>
<span id="cb1-687"><a href="#cb1-687" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">return</span> <span class="st">'wildtype'</span></span>
<span id="cb1-688"><a href="#cb1-688" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span>:</span>
<span id="cb1-689"><a href="#cb1-689" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">return</span> <span class="st">'homozygous mutant'</span></span>
<span id="cb1-690"><a href="#cb1-690" aria-hidden="true" tabindex="-1"></a>                <span class="cf">elif</span> <span class="bu">len</span>(_genotype) <span class="op">==</span>  <span class="dv">2</span>:</span>
<span id="cb1-691"><a href="#cb1-691" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">all</span>(_genotype.isin(mutant)):</span>
<span id="cb1-692"><a href="#cb1-692" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">return</span> <span class="st">'homozygous mutant'</span></span>
<span id="cb1-693"><a href="#cb1-693" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span>:</span>
<span id="cb1-694"><a href="#cb1-694" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">return</span> <span class="st">'heterozygous mutant'</span></span>
<span id="cb1-695"><a href="#cb1-695" aria-hidden="true" tabindex="-1"></a>                <span class="cf">elif</span> <span class="bu">len</span>(_genotype) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb1-696"><a href="#cb1-696" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">return</span> <span class="st">''</span></span>
<span id="cb1-697"><a href="#cb1-697" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb1-698"><a href="#cb1-698" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="st">'invaid genotype input!'</span>)</span>
<span id="cb1-699"><a href="#cb1-699" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-700"><a href="#cb1-700" aria-hidden="true" tabindex="-1"></a>             <span class="co"># df['phenotype'] = df.genotype.apply(lambda x: 'homozygous' if x in ['AA', "GG", "CC", "TT"] else 'heterozygous')</span></span>
<span id="cb1-701"><a href="#cb1-701" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'phenotype'</span>] <span class="op">=</span> df.<span class="bu">apply</span>(<span class="kw">lambda</span> row: _update_phenotype(row[<span class="st">'marker'</span>], row[<span class="st">'genotype'</span>]), axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb1-702"><a href="#cb1-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-703"><a href="#cb1-703" aria-hidden="true" tabindex="-1"></a>            <span class="co"># df['direction'] = '' # we dont know marker inforamtion in this file, need to be updated later</span></span>
<span id="cb1-704"><a href="#cb1-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-705"><a href="#cb1-705" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> pd.wide_to_long(df, stubnames<span class="op">=</span>[<span class="st">'base'</span>, <span class="st">'size'</span>, <span class="st">'height'</span>], </span>
<span id="cb1-706"><a href="#cb1-706" aria-hidden="true" tabindex="-1"></a>                                    j<span class="op">=</span><span class="st">'n_bases'</span>, i<span class="op">=</span>[<span class="st">'marker'</span>, <span class="st">'sample'</span>], sep<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb1-707"><a href="#cb1-707" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> df.reset_index()</span>
<span id="cb1-708"><a href="#cb1-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-709"><a href="#cb1-709" aria-hidden="true" tabindex="-1"></a>            <span class="kw">def</span> _color(marker, base):</span>
<span id="cb1-710"><a href="#cb1-710" aria-hidden="true" tabindex="-1"></a>                <span class="cf">return</span> marker.define_colors.get(base)</span>
<span id="cb1-711"><a href="#cb1-711" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb1-712"><a href="#cb1-712" aria-hidden="true" tabindex="-1"></a>            <span class="co"># def _direction(marker)</span></span>
<span id="cb1-713"><a href="#cb1-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-714"><a href="#cb1-714" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'color'</span>] <span class="op">=</span> df.<span class="bu">apply</span>(<span class="kw">lambda</span> row: _color(marker_info.get(row.marker), row.base), axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb1-715"><a href="#cb1-715" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'direction'</span>] <span class="op">=</span> df.<span class="bu">apply</span>(<span class="kw">lambda</span> row: marker_info.get(row.marker).direction, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb1-716"><a href="#cb1-716" aria-hidden="true" tabindex="-1"></a>            df[<span class="st">'label'</span>] <span class="op">=</span> df.<span class="bu">apply</span>(<span class="kw">lambda</span> row: marker_info.get(row.marker).marker_label, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb1-717"><a href="#cb1-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-718"><a href="#cb1-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-719"><a href="#cb1-719" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df</span>
<span id="cb1-720"><a href="#cb1-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-721"><a href="#cb1-721" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-722"><a href="#cb1-722" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> marker_table(<span class="va">self</span>, sample_name <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb1-723"><a href="#cb1-723" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Return a marker data information table</span></span>
<span id="cb1-724"><a href="#cb1-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-725"><a href="#cb1-725" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-726"><a href="#cb1-726" aria-hidden="true" tabindex="-1"></a><span class="co">            sample_name (str, optional): Filter marker table by sample name if required. Defaults to None.</span></span>
<span id="cb1-727"><a href="#cb1-727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-728"><a href="#cb1-728" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-729"><a href="#cb1-729" aria-hidden="true" tabindex="-1"></a><span class="co">            DataFrame: marker table</span></span>
<span id="cb1-730"><a href="#cb1-730" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-731"><a href="#cb1-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-732"><a href="#cb1-732" aria-hidden="true" tabindex="-1"></a>        cols <span class="op">=</span> [<span class="st">'sample'</span>, <span class="st">'gene'</span>, <span class="st">'marker'</span>, <span class="st">'label'</span>, <span class="st">'panel'</span>, <span class="st">'direction'</span>, <span class="st">'genotype'</span>, <span class="st">'phenotype'</span>]</span>
<span id="cb1-733"><a href="#cb1-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-734"><a href="#cb1-734" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> <span class="va">self</span>.clean_data[cols].drop_duplicates()</span>
<span id="cb1-735"><a href="#cb1-735" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-736"><a href="#cb1-736" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> sample_name:</span>
<span id="cb1-737"><a href="#cb1-737" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> df[df[<span class="st">'sample'</span>] <span class="op">==</span> sample_name]</span>
<span id="cb1-738"><a href="#cb1-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-739"><a href="#cb1-739" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df</span>
<span id="cb1-740"><a href="#cb1-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-741"><a href="#cb1-741" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-742"><a href="#cb1-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-743"><a href="#cb1-743" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> allele_table(<span class="va">self</span>, sample_name <span class="op">=</span> <span class="va">None</span>, called_filter<span class="op">=</span><span class="va">True</span>):</span>
<span id="cb1-744"><a href="#cb1-744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-745"><a href="#cb1-745" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Return a allele data information table. It can be different from bin file or fsa file.</span></span>
<span id="cb1-746"><a href="#cb1-746" aria-hidden="true" tabindex="-1"></a><span class="co">           Because some information we could not get from bin text file result and fsa file.</span></span>
<span id="cb1-747"><a href="#cb1-747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-748"><a href="#cb1-748" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb1-749"><a href="#cb1-749" aria-hidden="true" tabindex="-1"></a><span class="co">            _type_: _description_</span></span>
<span id="cb1-750"><a href="#cb1-750" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-751"><a href="#cb1-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-752"><a href="#cb1-752" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="va">self</span>.is_from_file:</span>
<span id="cb1-753"><a href="#cb1-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-754"><a href="#cb1-754" aria-hidden="true" tabindex="-1"></a>            cols <span class="op">=</span> [<span class="st">'sample'</span>, <span class="st">'gene'</span>, <span class="st">'marker'</span>, <span class="st">'label'</span>, <span class="st">'panel'</span>, <span class="st">'direction'</span>,</span>
<span id="cb1-755"><a href="#cb1-755" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'base'</span>, <span class="st">'size'</span>, <span class="st">'height'</span>, <span class="st">'color'</span>]</span>
<span id="cb1-756"><a href="#cb1-756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-757"><a href="#cb1-757" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb1-758"><a href="#cb1-758" aria-hidden="true" tabindex="-1"></a>            cols <span class="op">=</span> [<span class="st">'sample'</span>, <span class="st">'gene'</span>, <span class="st">'marker'</span>, <span class="st">'label'</span>, <span class="st">'panel'</span>, <span class="st">'direction'</span>,</span>
<span id="cb1-759"><a href="#cb1-759" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'base'</span>, <span class="st">'basetype'</span>, <span class="st">'min_bin'</span>, <span class="st">'max_bin'</span>, <span class="st">'min_height'</span>, </span>
<span id="cb1-760"><a href="#cb1-760" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'is_forward'</span>, <span class="st">'is_detected'</span>, <span class="st">'peak'</span>, <span class="st">'size'</span>, <span class="st">'height'</span>,</span>
<span id="cb1-761"><a href="#cb1-761" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'status'</span>, <span class="st">'message'</span>, <span class="st">'color'</span>]</span>
<span id="cb1-762"><a href="#cb1-762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-763"><a href="#cb1-763" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-764"><a href="#cb1-764" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> <span class="va">self</span>.clean_data[cols].drop_duplicates()</span>
<span id="cb1-765"><a href="#cb1-765" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-766"><a href="#cb1-766" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> sample_name:</span>
<span id="cb1-767"><a href="#cb1-767" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> df[df[<span class="st">'sample'</span>] <span class="op">==</span> sample_name]</span>
<span id="cb1-768"><a href="#cb1-768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-769"><a href="#cb1-769" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> called_filter:</span>
<span id="cb1-770"><a href="#cb1-770" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="va">self</span>.is_from_file:</span>
<span id="cb1-771"><a href="#cb1-771" aria-hidden="true" tabindex="-1"></a>                df <span class="op">=</span> df[pd.notna(df.color)] </span>
<span id="cb1-772"><a href="#cb1-772" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb1-773"><a href="#cb1-773" aria-hidden="true" tabindex="-1"></a>                df <span class="op">=</span> df[df.is_detected] </span>
<span id="cb1-774"><a href="#cb1-774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-775"><a href="#cb1-775" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df</span>
<span id="cb1-776"><a href="#cb1-776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-777"><a href="#cb1-777" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-778"><a href="#cb1-778" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> plotting_qc(<span class="va">self</span>, sample_name:<span class="bu">str</span>):</span>
<span id="cb1-779"><a href="#cb1-779" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Plot QC chromatogram based on called genotype: peak size, peak height.</span></span>
<span id="cb1-780"><a href="#cb1-780" aria-hidden="true" tabindex="-1"></a><span class="co">           It will plot all panel of this sample.</span></span>
<span id="cb1-781"><a href="#cb1-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-782"><a href="#cb1-782" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb1-783"><a href="#cb1-783" aria-hidden="true" tabindex="-1"></a><span class="co">            sample_name (str): name of sample</span></span>
<span id="cb1-784"><a href="#cb1-784" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb1-785"><a href="#cb1-785" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-786"><a href="#cb1-786" aria-hidden="true" tabindex="-1"></a>        allele_data <span class="op">=</span> <span class="va">self</span>.allele_table(sample_name)</span>
<span id="cb1-787"><a href="#cb1-787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-788"><a href="#cb1-788" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> panel <span class="kw">in</span> allele_data.panel.unique():</span>
<span id="cb1-789"><a href="#cb1-789" aria-hidden="true" tabindex="-1"></a>            plot_qc(allele_data[allele_data.panel <span class="op">==</span> panel])</span>
<span id="cb1-790"><a href="#cb1-790" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-791"><a href="#cb1-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-792"><a href="#cb1-792" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_intensity(size, height):</span>
<span id="cb1-793"><a href="#cb1-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-794"><a href="#cb1-794" aria-hidden="true" tabindex="-1"></a>    <span class="co">'''Generate intensity base on height of each allele'''</span></span>
<span id="cb1-795"><a href="#cb1-795" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-796"><a href="#cb1-796" aria-hidden="true" tabindex="-1"></a>    <span class="co"># generate width based on height</span></span>
<span id="cb1-797"><a href="#cb1-797" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-798"><a href="#cb1-798" aria-hidden="true" tabindex="-1"></a>    _range <span class="op">=</span> <span class="fl">2.5</span> <span class="cf">if</span> height <span class="op">&lt;</span> <span class="dv">4000</span> <span class="cf">else</span> <span class="fl">2.7</span></span>
<span id="cb1-799"><a href="#cb1-799" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-800"><a href="#cb1-800" aria-hidden="true" tabindex="-1"></a>    width <span class="op">=</span> np.arange(<span class="op">-</span><span class="dv">1</span><span class="op">*</span>_range,_range,<span class="fl">0.01</span>)</span>
<span id="cb1-801"><a href="#cb1-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-802"><a href="#cb1-802" aria-hidden="true" tabindex="-1"></a>    intensity <span class="op">=</span> height<span class="op">*</span>np.exp(<span class="op">-</span><span class="dv">15</span><span class="op">*</span>np.log(<span class="dv">1</span><span class="op">+</span>(width<span class="op">**</span><span class="dv">2</span>)<span class="op">/</span><span class="dv">10</span>))</span>
<span id="cb1-803"><a href="#cb1-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-804"><a href="#cb1-804" aria-hidden="true" tabindex="-1"></a>    <span class="co"># set 1 first and last intensity = 0 to make sure all all k</span></span>
<span id="cb1-805"><a href="#cb1-805" aria-hidden="true" tabindex="-1"></a>    intensity[<span class="dv">0</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb1-806"><a href="#cb1-806" aria-hidden="true" tabindex="-1"></a>    intensity[<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb1-807"><a href="#cb1-807" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-808"><a href="#cb1-808" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> width, intensity</span>
<span id="cb1-809"><a href="#cb1-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-810"><a href="#cb1-810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-811"><a href="#cb1-811" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_qc(data, xlim <span class="op">=</span>(<span class="dv">20</span>,<span class="dv">80</span>), </span>
<span id="cb1-812"><a href="#cb1-812" aria-hidden="true" tabindex="-1"></a>         figsize <span class="op">=</span>(<span class="dv">15</span>,<span class="dv">4</span>),</span>
<span id="cb1-813"><a href="#cb1-813" aria-hidden="true" tabindex="-1"></a>         dpi <span class="op">=</span> <span class="dv">200</span>,</span>
<span id="cb1-814"><a href="#cb1-814" aria-hidden="true" tabindex="-1"></a>         alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span id="cb1-815"><a href="#cb1-815" aria-hidden="true" tabindex="-1"></a>         marker_labels<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb1-816"><a href="#cb1-816" aria-hidden="true" tabindex="-1"></a>         ex_height <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span id="cb1-817"><a href="#cb1-817" aria-hidden="true" tabindex="-1"></a>         ex_annotation <span class="op">=</span> <span class="fl">0.05</span>):</span>
<span id="cb1-818"><a href="#cb1-818" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-819"><a href="#cb1-819" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> data.sort_values(<span class="st">'height'</span>)</span>
<span id="cb1-820"><a href="#cb1-820" aria-hidden="true" tabindex="-1"></a>    max_height <span class="op">=</span> np.<span class="bu">max</span>(data.height)</span>
<span id="cb1-821"><a href="#cb1-821" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-822"><a href="#cb1-822" aria-hidden="true" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>figsize, dpi<span class="op">=</span>dpi)</span>
<span id="cb1-823"><a href="#cb1-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-824"><a href="#cb1-824" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="op">*</span>xlim)</span>
<span id="cb1-825"><a href="#cb1-825" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="dv">0</span>, max_height <span class="op">+</span>  max_height <span class="op">*</span> ex_height)</span>
<span id="cb1-826"><a href="#cb1-826" aria-hidden="true" tabindex="-1"></a>    ax.grid(linestyle<span class="op">=</span><span class="st">'--'</span>, alpha <span class="op">=</span> <span class="fl">0.3</span>, color <span class="op">=</span><span class="st">'gray'</span>, linewidth <span class="op">=</span> <span class="fl">0.3</span>)</span>
<span id="cb1-827"><a href="#cb1-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-828"><a href="#cb1-828" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> idx, row <span class="kw">in</span> data.iterrows():</span>
<span id="cb1-829"><a href="#cb1-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-830"><a href="#cb1-830" aria-hidden="true" tabindex="-1"></a>        width, y <span class="op">=</span> generate_intensity(row[<span class="st">'size'</span>], row.height)</span>
<span id="cb1-831"><a href="#cb1-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-832"><a href="#cb1-832" aria-hidden="true" tabindex="-1"></a>        x <span class="op">=</span> width <span class="op">+</span> row[<span class="st">'size'</span>]</span>
<span id="cb1-833"><a href="#cb1-833" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(row.size)</span></span>
<span id="cb1-834"><a href="#cb1-834" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-835"><a href="#cb1-835" aria-hidden="true" tabindex="-1"></a>        ax.fill(x, y, color<span class="op">=</span>row.color, alpha<span class="op">=</span>alpha)</span>
<span id="cb1-836"><a href="#cb1-836" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-837"><a href="#cb1-837" aria-hidden="true" tabindex="-1"></a>    <span class="co"># process marker information</span></span>
<span id="cb1-838"><a href="#cb1-838" aria-hidden="true" tabindex="-1"></a>    props <span class="op">=</span> <span class="bu">dict</span>(boxstyle<span class="op">=</span><span class="st">'round'</span>, facecolor<span class="op">=</span><span class="st">'#f8f9fa'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb1-839"><a href="#cb1-839" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> marker_labels:</span>
<span id="cb1-840"><a href="#cb1-840" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> data.groupby([<span class="st">'label'</span>], as_index<span class="op">=</span><span class="va">False</span>).agg({<span class="st">'size'</span>:np.average, <span class="st">'height'</span>:np.<span class="bu">max</span>})</span>
<span id="cb1-841"><a href="#cb1-841" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-842"><a href="#cb1-842" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> _, row <span class="kw">in</span> tmp.iterrows():</span>
<span id="cb1-843"><a href="#cb1-843" aria-hidden="true" tabindex="-1"></a>            ax.text(row[<span class="st">'size'</span>],</span>
<span id="cb1-844"><a href="#cb1-844" aria-hidden="true" tabindex="-1"></a>                    row.height <span class="op">+</span> max_height <span class="op">*</span> ex_annotation,</span>
<span id="cb1-845"><a href="#cb1-845" aria-hidden="true" tabindex="-1"></a>                    row.label, </span>
<span id="cb1-846"><a href="#cb1-846" aria-hidden="true" tabindex="-1"></a>                    ha<span class="op">=</span><span class="st">'center'</span>,</span>
<span id="cb1-847"><a href="#cb1-847" aria-hidden="true" tabindex="-1"></a>                    fontsize <span class="op">=</span> <span class="dv">11</span>,</span>
<span id="cb1-848"><a href="#cb1-848" aria-hidden="true" tabindex="-1"></a>                    bbox<span class="op">=</span>props)</span>
<span id="cb1-849"><a href="#cb1-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-850"><a href="#cb1-850" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb1-851"><a href="#cb1-851" aria-hidden="true" tabindex="-1"></a>    plt.show()     </span>
<span id="cb1-852"><a href="#cb1-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-853"><a href="#cb1-853" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-854"><a href="#cb1-854" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-855"><a href="#cb1-855" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> call_from_fsa(fsa_files:<span class="bu">list</span>, target_markers:<span class="bu">dict</span>, reference:<span class="bu">list</span>, val_plot<span class="op">=</span><span class="va">True</span>, intensity_plot <span class="op">=</span> <span class="va">True</span>):</span>
<span id="cb1-856"><a href="#cb1-856" aria-hidden="true" tabindex="-1"></a>    <span class="co"># load reference size</span></span>
<span id="cb1-857"><a href="#cb1-857" aria-hidden="true" tabindex="-1"></a>    ref <span class="op">=</span> Reference(reference)</span>
<span id="cb1-858"><a href="#cb1-858" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-859"><a href="#cb1-859" aria-hidden="true" tabindex="-1"></a>    markers <span class="op">=</span> target_markers.copy()</span>
<span id="cb1-860"><a href="#cb1-860" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print_marker(markers)</span></span>
<span id="cb1-861"><a href="#cb1-861" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-862"><a href="#cb1-862" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> fsa_file <span class="kw">in</span> fsa_files:</span>
<span id="cb1-863"><a href="#cb1-863" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-864"><a href="#cb1-864" aria-hidden="true" tabindex="-1"></a>        fsa <span class="op">=</span> FSA(fsa_file)</span>
<span id="cb1-865"><a href="#cb1-865" aria-hidden="true" tabindex="-1"></a>        ref_intensity <span class="op">=</span> fsa.reference_intensity</span>
<span id="cb1-866"><a href="#cb1-866" aria-hidden="true" tabindex="-1"></a>        ref_peaks, ref_peaks_height <span class="op">=</span> finding_peaks(ref_intensity, min_height<span class="op">=</span><span class="dv">800</span>)</span>
<span id="cb1-867"><a href="#cb1-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-868"><a href="#cb1-868" aria-hidden="true" tabindex="-1"></a>        <span class="co"># build model</span></span>
<span id="cb1-869"><a href="#cb1-869" aria-hidden="true" tabindex="-1"></a>        model <span class="op">=</span> LeastSquared(ref_peaks, ref.size, degree<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb1-870"><a href="#cb1-870" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-871"><a href="#cb1-871" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> val_plot:</span>
<span id="cb1-872"><a href="#cb1-872" aria-hidden="true" tabindex="-1"></a>            model.val_plot(fsa.name)</span>
<span id="cb1-873"><a href="#cb1-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-874"><a href="#cb1-874" aria-hidden="true" tabindex="-1"></a>        <span class="co"># predict base from model</span></span>
<span id="cb1-875"><a href="#cb1-875" aria-hidden="true" tabindex="-1"></a>        pred <span class="op">=</span> model.predict(<span class="bu">range</span>(fsa.chanel_size))</span>
<span id="cb1-876"><a href="#cb1-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-877"><a href="#cb1-877" aria-hidden="true" tabindex="-1"></a>        <span class="co"># extract interested range 0 to 200 bp</span></span>
<span id="cb1-878"><a href="#cb1-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-879"><a href="#cb1-879" aria-hidden="true" tabindex="-1"></a>        min_index, max_index <span class="op">=</span> get_index_by_base_range(pred, base_ranges<span class="op">=</span>(<span class="dv">0</span>,<span class="dv">200</span>))</span>
<span id="cb1-880"><a href="#cb1-880" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-881"><a href="#cb1-881" aria-hidden="true" tabindex="-1"></a>        select_base <span class="op">=</span> pred[min_index:max_index]</span>
<span id="cb1-882"><a href="#cb1-882" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-883"><a href="#cb1-883" aria-hidden="true" tabindex="-1"></a>        <span class="co"># extract panel</span></span>
<span id="cb1-884"><a href="#cb1-884" aria-hidden="true" tabindex="-1"></a>        fsa_panel <span class="op">=</span> fsa.name.split(<span class="st">'-'</span>)[<span class="dv">0</span>]</span>
<span id="cb1-885"><a href="#cb1-885" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-886"><a href="#cb1-886" aria-hidden="true" tabindex="-1"></a>        <span class="co"># find peak and update results to markers</span></span>
<span id="cb1-887"><a href="#cb1-887" aria-hidden="true" tabindex="-1"></a>        finding_peak_by_markers(markers,fsa, fsa_panel,select_base, (min_index, max_index))</span>
<span id="cb1-888"><a href="#cb1-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-889"><a href="#cb1-889" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> intensity_plot:</span>
<span id="cb1-890"><a href="#cb1-890" aria-hidden="true" tabindex="-1"></a>            plot_raw_intensity(fsa, min_index, max_index, select_base, (<span class="dv">0</span>,<span class="dv">130</span>))</span>
<span id="cb1-891"><a href="#cb1-891" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-892"><a href="#cb1-892" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print_marker(markers)</span></span>
<span id="cb1-893"><a href="#cb1-893" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-894"><a href="#cb1-894" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> markers</span>
<span id="cb1-895"><a href="#cb1-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-896"><a href="#cb1-896" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-897"><a href="#cb1-897" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Definition:</span>
<span id="cb1-898"><a href="#cb1-898" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""This class is to load the tables data from definition file the already processed before.</span></span>
<span id="cb1-899"><a href="#cb1-899" aria-hidden="true" tabindex="-1"></a><span class="co">       There are 3 tables in this tables file and they are interconnecting each other.</span></span>
<span id="cb1-900"><a href="#cb1-900" aria-hidden="true" tabindex="-1"></a><span class="co">       This class will help us to quick extract the information from defintion tables</span></span>
<span id="cb1-901"><a href="#cb1-901" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-902"><a href="#cb1-902" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, tables) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-903"><a href="#cb1-903" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.tables <span class="op">=</span> tables</span>
<span id="cb1-904"><a href="#cb1-904" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.gene_table <span class="op">=</span> <span class="va">self</span>.tables[<span class="st">'gene_table'</span>]</span>
<span id="cb1-905"><a href="#cb1-905" aria-hidden="true" tabindex="-1"></a>        <span class="co"># self.marker_table = self.tables['marker_table']</span></span>
<span id="cb1-906"><a href="#cb1-906" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.genotype_marker_table <span class="op">=</span> <span class="va">self</span>.tables[<span class="st">'genotype_marker_table'</span>]</span>
<span id="cb1-907"><a href="#cb1-907" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.marker_table <span class="op">=</span> <span class="va">self</span>._add_gene_to_marker_table()</span>
<span id="cb1-908"><a href="#cb1-908" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-909"><a href="#cb1-909" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-910"><a href="#cb1-910" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> extract_geneid(<span class="va">self</span>, gene):</span>
<span id="cb1-911"><a href="#cb1-911" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.gene_table[<span class="va">self</span>.gene_table[<span class="st">'gene'</span>] <span class="op">==</span> gene][<span class="st">'gene_id'</span>]</span>
<span id="cb1-912"><a href="#cb1-912" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-913"><a href="#cb1-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-914"><a href="#cb1-914" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> _add_gene_to_marker_table(<span class="va">self</span>):</span>
<span id="cb1-915"><a href="#cb1-915" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.tables[<span class="st">'marker_table'</span>].merge(<span class="va">self</span>.gene_table, how <span class="op">=</span><span class="st">'left'</span>, on<span class="op">=</span><span class="st">'gene_id'</span>)</span>
<span id="cb1-916"><a href="#cb1-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-917"><a href="#cb1-917" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-918"><a href="#cb1-918" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> extract_marker(<span class="va">self</span>, gene<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-919"><a href="#cb1-919" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> gene:</span>
<span id="cb1-920"><a href="#cb1-920" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="va">self</span>.marker_table[<span class="va">self</span>.marker_table[<span class="st">'gene'</span>] <span class="op">==</span> gene]</span>
<span id="cb1-921"><a href="#cb1-921" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb1-922"><a href="#cb1-922" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="va">self</span>.marker_table</span>
<span id="cb1-923"><a href="#cb1-923" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-924"><a href="#cb1-924" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-925"><a href="#cb1-925" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_target_marker(<span class="va">self</span>, gene:<span class="bu">str</span>):</span>
<span id="cb1-926"><a href="#cb1-926" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-927"><a href="#cb1-927" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> <span class="va">self</span>.genotype_marker_table</span>
<span id="cb1-928"><a href="#cb1-928" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> gene:</span>
<span id="cb1-929"><a href="#cb1-929" aria-hidden="true" tabindex="-1"></a>            df <span class="op">=</span> <span class="va">self</span>.genotype_marker_table[<span class="va">self</span>.genotype_marker_table[<span class="st">'gene'</span>] <span class="op">==</span> gene]</span>
<span id="cb1-930"><a href="#cb1-930" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb1-931"><a href="#cb1-931" aria-hidden="true" tabindex="-1"></a>        genotype_id <span class="op">=</span> df[<span class="st">'genotype_id'</span>].tolist()[<span class="dv">0</span>]</span>
<span id="cb1-932"><a href="#cb1-932" aria-hidden="true" tabindex="-1"></a>        marker_names <span class="op">=</span> df[df[<span class="st">'genotype_id'</span>] <span class="op">==</span> genotype_id][<span class="st">'marker'</span>]</span>
<span id="cb1-933"><a href="#cb1-933" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-934"><a href="#cb1-934" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> df[[<span class="st">'gene'</span>, <span class="st">'genotype'</span>, <span class="st">'marker'</span>, <span class="st">'value'</span>]].pivot(index<span class="op">=</span><span class="st">'genotype'</span>, columns<span class="op">=</span><span class="st">'marker'</span>, values<span class="op">=</span><span class="st">'value'</span>)</span>
<span id="cb1-935"><a href="#cb1-935" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> df[marker_names]</span>
<span id="cb1-936"><a href="#cb1-936" aria-hidden="true" tabindex="-1"></a>        df[<span class="st">'target'</span>] <span class="op">=</span> df[marker_names].<span class="bu">apply</span>(<span class="kw">lambda</span> row: <span class="st">'_'</span>.join(row.values.astype(<span class="bu">str</span>)), axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb1-937"><a href="#cb1-937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-938"><a href="#cb1-938" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> {<span class="st">'marker_name'</span>:marker_names, <span class="st">'pattern'</span>:df}</span>
<span id="cb1-939"><a href="#cb1-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-940"><a href="#cb1-940" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-941"><a href="#cb1-941" aria-hidden="true" tabindex="-1"></a><span class="co"># def check_if_marker_called(defined_markers, sample_markers):</span></span>
<span id="cb1-942"><a href="#cb1-942" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-943"><a href="#cb1-943" aria-hidden="true" tabindex="-1"></a><span class="co">#     df = defined_markers.merge(sample_markers, how='left', on='marker', suffixes=('_defined', ''))</span></span>
<span id="cb1-944"><a href="#cb1-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-945"><a href="#cb1-945" aria-hidden="true" tabindex="-1"></a><span class="co">#     return df.marker[np.where(pd.isna(df.genotype))[0]]</span></span>
<span id="cb1-946"><a href="#cb1-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-947"><a href="#cb1-947" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-948"><a href="#cb1-948" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_sample_pattern(sample_markers, target_marker_names):</span>
<span id="cb1-949"><a href="#cb1-949" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Generate comparison pattern of sample to compare with patterns from definition tables.clear</span></span>
<span id="cb1-950"><a href="#cb1-950" aria-hidden="true" tabindex="-1"></a><span class="co">       Example: </span></span>
<span id="cb1-951"><a href="#cb1-951" aria-hidden="true" tabindex="-1"></a><span class="co">       Marker_01    AA</span></span>
<span id="cb1-952"><a href="#cb1-952" aria-hidden="true" tabindex="-1"></a><span class="co">       Marker_02    CT</span></span>
<span id="cb1-953"><a href="#cb1-953" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-954"><a href="#cb1-954" aria-hidden="true" tabindex="-1"></a><span class="co">       Pattern: AA_(CT|TC)</span></span>
<span id="cb1-955"><a href="#cb1-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-956"><a href="#cb1-956" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-957"><a href="#cb1-957" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-958"><a href="#cb1-958" aria-hidden="true" tabindex="-1"></a><span class="co">        sample_markers (DataFrame): A DataFrame contains all called genotypes</span></span>
<span id="cb1-959"><a href="#cb1-959" aria-hidden="true" tabindex="-1"></a><span class="co">        target_marker_names (list): A list of the target marker name from definition tables</span></span>
<span id="cb1-960"><a href="#cb1-960" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-961"><a href="#cb1-961" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-962"><a href="#cb1-962" aria-hidden="true" tabindex="-1"></a><span class="co">        str: the generated pattern</span></span>
<span id="cb1-963"><a href="#cb1-963" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-964"><a href="#cb1-964" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> sample_markers[sample_markers.marker.isin(target_marker_names)]</span>
<span id="cb1-965"><a href="#cb1-965" aria-hidden="true" tabindex="-1"></a>    cols <span class="op">=</span> [<span class="st">'sample'</span>,<span class="st">'gene'</span>, <span class="st">'marker'</span>, <span class="st">'genotype'</span>] </span>
<span id="cb1-966"><a href="#cb1-966" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df[cols]</span>
<span id="cb1-967"><a href="#cb1-967" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-968"><a href="#cb1-968" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> _generate_pattern(genotype):</span>
<span id="cb1-969"><a href="#cb1-969" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-970"><a href="#cb1-970" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> genotype <span class="kw">not</span> <span class="kw">in</span> [<span class="st">'AA'</span>, <span class="st">'GG'</span>, <span class="st">'CC'</span>, <span class="st">'TT'</span>]:</span>
<span id="cb1-971"><a href="#cb1-971" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="ss">f'(</span><span class="sc">{</span>genotype<span class="sc">}</span><span class="ss">|</span><span class="sc">{</span>genotype[::<span class="op">-</span><span class="dv">1</span>]<span class="sc">}</span><span class="ss">)'</span></span>
<span id="cb1-972"><a href="#cb1-972" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb1-973"><a href="#cb1-973" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> genotype</span>
<span id="cb1-974"><a href="#cb1-974" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-975"><a href="#cb1-975" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'genotype'</span>] <span class="op">=</span> df[<span class="st">'genotype'</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: _generate_pattern(x))</span>
<span id="cb1-976"><a href="#cb1-976" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-977"><a href="#cb1-977" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.pivot(index<span class="op">=</span><span class="st">'sample'</span>, columns<span class="op">=</span><span class="st">'marker'</span>, values<span class="op">=</span><span class="st">'genotype'</span>)</span>
<span id="cb1-978"><a href="#cb1-978" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df[target_marker_names]</span>
<span id="cb1-979"><a href="#cb1-979" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'exp_sample'</span>] <span class="op">=</span> df[target_marker_names].<span class="bu">apply</span>(<span class="kw">lambda</span> row: <span class="st">'_'</span>.join(row.values.astype(<span class="bu">str</span>)), axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb1-980"><a href="#cb1-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-981"><a href="#cb1-981" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df[<span class="st">'exp_sample'</span>][<span class="dv">0</span>]</span>
<span id="cb1-982"><a href="#cb1-982" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-983"><a href="#cb1-983" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-984"><a href="#cb1-984" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> call_star_allele(sample_pattern, targets):</span>
<span id="cb1-985"><a href="#cb1-985" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Matching pattern of sample and defintion table</span></span>
<span id="cb1-986"><a href="#cb1-986" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-987"><a href="#cb1-987" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-988"><a href="#cb1-988" aria-hidden="true" tabindex="-1"></a><span class="co">        sample_pattern (str): sample pattern</span></span>
<span id="cb1-989"><a href="#cb1-989" aria-hidden="true" tabindex="-1"></a><span class="co">        targets (DataFrame): A data frame includes all generate pattern for all combination star allele (diplotype) from definition table.</span></span>
<span id="cb1-990"><a href="#cb1-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-991"><a href="#cb1-991" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb1-992"><a href="#cb1-992" aria-hidden="true" tabindex="-1"></a><span class="co">        (str):  All called diplotype (star allele combintations) that can match with this sample.</span></span>
<span id="cb1-993"><a href="#cb1-993" aria-hidden="true" tabindex="-1"></a><span class="co">                If more than one matched, the result will be joined and separeted by "|". For example: *1/*2xN|*1xN/*2</span></span>
<span id="cb1-994"><a href="#cb1-994" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-995"><a href="#cb1-995" aria-hidden="true" tabindex="-1"></a>    matched <span class="op">=</span> [<span class="bu">bool</span>(re.fullmatch(pattern<span class="op">=</span>sample_pattern, string<span class="op">=</span>target))</span>
<span id="cb1-996"><a href="#cb1-996" aria-hidden="true" tabindex="-1"></a>                       <span class="cf">for</span> target <span class="kw">in</span> targets[<span class="st">'target'</span>]]</span>
<span id="cb1-997"><a href="#cb1-997" aria-hidden="true" tabindex="-1"></a>    genotype <span class="op">=</span> <span class="st">'|'</span>.join(targets.iloc[np.where(matched)].index.values.tolist())</span>
<span id="cb1-998"><a href="#cb1-998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-999"><a href="#cb1-999" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> genotype</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-defined-marker-with-colors" class="level2">
<h2 class="anchored" data-anchor-id="load-defined-marker-with-colors">Load defined marker with colors</h2>
<p>This form is test form and it was based on the format from PGx team.</p>
<p>However this form have a disavantage if testing allele is not biallelic for example triallelic (A &gt; C/T). and this requires a new format to overcome this issue.</p>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>peak_table <span class="op">=</span> pd.read_excel(<span class="st">'./resource/test_bin_peak.xlsx'</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>defined_markers <span class="op">=</span> generate_markers(peak_table)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>liz120 <span class="op">=</span> [<span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">25</span>, <span class="dv">35</span>, <span class="dv">50</span>, <span class="dv">62</span>, <span class="dv">80</span>, <span class="dv">110</span>, <span class="dv">120</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-call-from-fsa-file-with-cyp2d6-and-2-panels" class="level2">
<h2 class="anchored" data-anchor-id="example-call-from-fsa-file-with-cyp2d6-and-2-panels">Example call from fsa file with CYP2D6 and 2 panels</h2>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>fsa_files <span class="op">=</span> [<span class="st">'./raw_data/fsa/PA180404A/S1-20180328-230857_G02_SNaPshot50_POP7_E5_2018-04-06-16-40-43.fsa'</span>,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>             <span class="st">'./raw_data/fsa/PA180404A/S2-20180328-230857_H02_SNaPshot50_POP7_E5_2018-04-06-16-40-43.fsa'</span>]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>liz120 <span class="op">=</span> [<span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">25</span>, <span class="dv">35</span>, <span class="dv">50</span>, <span class="dv">62</span>, <span class="dv">80</span>, <span class="dv">110</span>, <span class="dv">120</span>]</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'./resource/tables.pdata'</span>, <span class="st">'rb'</span>) <span class="im">as</span> t:</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    tables <span class="op">=</span> pickle.load(t)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>definition <span class="op">=</span> Definition(tables)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co">#load origin markers</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>s1 <span class="op">=</span> call_from_fsa(fsa_files, defined_markers, liz120)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># call_from_fsa(fsa_files, defined_markers, liz120)</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># input:</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 1. fsa_files</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 2. defined_markers = generate_markers</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Input: (1): test_bin_peak.xlsx</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 3. liz120: ref size: e.g.: [15, 20, 25, 35, 50, 62, 80, 110, 120]</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># print(s1)</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>gene <span class="op">=</span> <span class="st">'CYP2D6'</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> GenotypeResult()</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>g.from_fsa_call(s1)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> g.sample_list:</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>    sample_data <span class="op">=</span> g.marker_table(sample)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>    targets <span class="op">=</span> definition.get_target_marker(gene)</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>    sample_pattern <span class="op">=</span> generate_sample_pattern(sample_data, targets[<span class="st">'marker_name'</span>])</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    result <span class="op">=</span> call_star_allele(sample_pattern, targets[<span class="st">'pattern'</span>])</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>sample<span class="sc">}</span><span class="ss">:</span><span class="sc">{</span>result<span class="sc">}</span><span class="ss">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-4-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-4-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-4-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-4-output-4.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>20180328-230857:*1/*10B</code></pre>
</div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>g.plotting_qc(<span class="st">'20180328-230857'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-5-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-5-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="example-call-from-bin-file" class="level2">
<h2 class="anchored" data-anchor-id="example-call-from-bin-file">Example call from bin file</h2>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>g2 <span class="op">=</span> GenotypeResult()</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>g2.from_bin_file(<span class="st">'./raw_data/bin_text/PA180404A_CYP2D6_Example.txt'</span>, defined_markers)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>gene <span class="op">=</span> <span class="st">'CYP2D6'</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> g2.sample_list:</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    sample_data <span class="op">=</span> g2.marker_table(sample)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># defined_markers = definition.extract_marker(gene)</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    targets <span class="op">=</span> definition.get_target_marker(gene)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    sample_pattern <span class="op">=</span> generate_sample_pattern(sample_data, targets[<span class="st">'marker_name'</span>])</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    result <span class="op">=</span> call_star_allele(sample_pattern, targets[<span class="st">'pattern'</span>])</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>sample<span class="sc">}</span><span class="ss">:</span><span class="sc">{</span>result<span class="sc">}</span><span class="ss">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>20180328-230857:*1/*10B
PTC:*1/*1
NTC:</code></pre>
</div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>g2.marker_table()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample</th>
      <th>gene</th>
      <th>marker</th>
      <th>label</th>
      <th>panel</th>
      <th>direction</th>
      <th>genotype</th>
      <th>phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>CT</td>
      <td>heterozygous mutant</td>
    </tr>
    <tr>
      <th>4</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_003</td>
      <td>CYP2D6_49</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>TT</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>6</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_004</td>
      <td>CYP2D6_21</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>8</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_005</td>
      <td>CYP2D6_41</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>10</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_006</td>
      <td>CYP2D6_52</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>12</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_007</td>
      <td>CYP2D6_18</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>14</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_008</td>
      <td>CYP2D6_2</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>CC</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>16</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_009</td>
      <td>CYP2D6_60</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>18</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_010</td>
      <td>CYP2D6_5</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>20</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_011</td>
      <td>CYP2D6_4</td>
      <td>S2</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>22</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_013</td>
      <td>CYP2D6_17</td>
      <td>S2</td>
      <td>Forward</td>
      <td>CC</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>24</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_012</td>
      <td>CYP2D6_3</td>
      <td>S2</td>
      <td>Forward</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>26</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_015</td>
      <td>CYP2D6_6</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>TT</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>28</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_014</td>
      <td>CYP2D6_9</td>
      <td>S2</td>
      <td>Forward</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>30</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_016</td>
      <td>CYP2D6_29</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>32</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_017</td>
      <td>CYP2D6_XN</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>34</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>36</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>CC</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>38</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_003</td>
      <td>CYP2D6_49</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>TT</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>40</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_004</td>
      <td>CYP2D6_21</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>42</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_005</td>
      <td>CYP2D6_41</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>44</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_006</td>
      <td>CYP2D6_52</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>46</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_007</td>
      <td>CYP2D6_18</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>48</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_008</td>
      <td>CYP2D6_2</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>CC</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>50</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_009</td>
      <td>CYP2D6_60</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>52</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_010</td>
      <td>CYP2D6_5</td>
      <td>S1</td>
      <td>Reverse</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>54</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_011</td>
      <td>CYP2D6_4</td>
      <td>S2</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>56</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_013</td>
      <td>CYP2D6_17</td>
      <td>S2</td>
      <td>Forward</td>
      <td>CC</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>58</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_012</td>
      <td>CYP2D6_3</td>
      <td>S2</td>
      <td>Forward</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>60</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_015</td>
      <td>CYP2D6_6</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>TT</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>62</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_014</td>
      <td>CYP2D6_9</td>
      <td>S2</td>
      <td>Forward</td>
      <td>AA</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>64</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_016</td>
      <td>CYP2D6_29</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>66</th>
      <td>PTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_017</td>
      <td>CYP2D6_XN</td>
      <td>S2</td>
      <td>Reverse</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
    <tr>
      <th>68</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>70</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>72</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_003</td>
      <td>CYP2D6_49</td>
      <td>S1</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>74</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_004</td>
      <td>CYP2D6_21</td>
      <td>S1</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>76</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_005</td>
      <td>CYP2D6_41</td>
      <td>S1</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>78</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_006</td>
      <td>CYP2D6_52</td>
      <td>S1</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>80</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_007</td>
      <td>CYP2D6_18</td>
      <td>S1</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>82</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_008</td>
      <td>CYP2D6_2</td>
      <td>S1</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>84</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_009</td>
      <td>CYP2D6_60</td>
      <td>S1</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>86</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_010</td>
      <td>CYP2D6_5</td>
      <td>S1</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>88</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_011</td>
      <td>CYP2D6_4</td>
      <td>S2</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>90</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_013</td>
      <td>CYP2D6_17</td>
      <td>S2</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>92</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_012</td>
      <td>CYP2D6_3</td>
      <td>S2</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>94</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_015</td>
      <td>CYP2D6_6</td>
      <td>S2</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>96</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_014</td>
      <td>CYP2D6_9</td>
      <td>S2</td>
      <td>Forward</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>98</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_016</td>
      <td>CYP2D6_29</td>
      <td>S2</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>100</th>
      <td>NTC</td>
      <td>CYP2D6</td>
      <td>CYP2D6_017</td>
      <td>CYP2D6_XN</td>
      <td>S2</td>
      <td>Reverse</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>g2.marker_table().head(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample</th>
      <th>gene</th>
      <th>marker</th>
      <th>label</th>
      <th>panel</th>
      <th>direction</th>
      <th>genotype</th>
      <th>phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>GG</td>
      <td>wildtype</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>g2.allele_table(sample_name<span class="op">=</span><span class="st">'20180328-230857'</span>, called_filter<span class="op">=</span><span class="va">True</span>).head(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample</th>
      <th>gene</th>
      <th>marker</th>
      <th>label</th>
      <th>panel</th>
      <th>direction</th>
      <th>base</th>
      <th>size</th>
      <th>height</th>
      <th>color</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>G</td>
      <td>29.04</td>
      <td>6967.0</td>
      <td>blue</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>g2.plotting_qc(<span class="st">'20180328-230857'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-10-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-10-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>g2.plotting_qc(<span class="st">'PTC'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-11-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-11-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="update-bin-range-to-call-genotype-example" class="level1">
<h1>Update bin range to call genotype example</h1>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>fsa_files <span class="op">=</span> [<span class="st">'./raw_data/fsa/PA180404A/S1-20180328-230857_G02_SNaPshot50_POP7_E5_2018-04-06-16-40-43.fsa'</span>,</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>             <span class="st">'./raw_data/fsa/PA180404A/S2-20180328-230857_H02_SNaPshot50_POP7_E5_2018-04-06-16-40-43.fsa'</span>]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>liz120 <span class="op">=</span> [<span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">25</span>, <span class="dv">35</span>, <span class="dv">50</span>, <span class="dv">62</span>, <span class="dv">80</span>, <span class="dv">110</span>, <span class="dv">120</span>]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'./resource/tables.pdata'</span>, <span class="st">'rb'</span>) <span class="im">as</span> t:</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    tables <span class="op">=</span> pickle.load(t)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>definition <span class="op">=</span> Definition(tables)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>s3_markers <span class="op">=</span> generate_markers(peak_table)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>s3_markers[<span class="st">'CYP2D6_001'</span>].update_allele_config(allele_number <span class="op">=</span> <span class="dv">0</span>, min_bin <span class="op">=</span> <span class="dv">20</span>, max_bin <span class="op">=</span> <span class="dv">25</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>s3 <span class="op">=</span> call_from_fsa(fsa_files, s3_markers, liz120, val_plot<span class="op">=</span><span class="va">False</span>, intensity_plot<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>g3 <span class="op">=</span> GenotypeResult()</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>g3.from_fsa_call(s3)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>g3.allele_table(called_filter<span class="op">=</span><span class="va">False</span>).head(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample</th>
      <th>gene</th>
      <th>marker</th>
      <th>label</th>
      <th>panel</th>
      <th>direction</th>
      <th>base</th>
      <th>basetype</th>
      <th>min_bin</th>
      <th>max_bin</th>
      <th>min_height</th>
      <th>is_forward</th>
      <th>is_detected</th>
      <th>peak</th>
      <th>size</th>
      <th>height</th>
      <th>status</th>
      <th>message</th>
      <th>color</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>G</td>
      <td>wildtype</td>
      <td>20</td>
      <td>25</td>
      <td>500</td>
      <td>1</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>Peak(s) could not be detected. Please check pe...</td>
      <td>blue</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>A</td>
      <td>mutant</td>
      <td>27</td>
      <td>36</td>
      <td>500</td>
      <td>1</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>Peak(s) could not be detected. Please check pe...</td>
      <td>green</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>C</td>
      <td>wildtype</td>
      <td>28</td>
      <td>38</td>
      <td>500</td>
      <td>1</td>
      <td>True</td>
      <td>32</td>
      <td>32.6</td>
      <td>901.0</td>
      <td>ok</td>
      <td></td>
      <td>black</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20180328-230857</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>T</td>
      <td>mutant</td>
      <td>31</td>
      <td>37</td>
      <td>500</td>
      <td>1</td>
      <td>True</td>
      <td>32</td>
      <td>35.55</td>
      <td>1019.0</td>
      <td>ok</td>
      <td></td>
      <td>red</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># qc plot</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>g3.plotting_qc(<span class="st">'20180328-230857'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-13-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-13-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>s3_markers[<span class="st">'CYP2D6_001'</span>].update_allele_config(allele_number <span class="op">=</span> <span class="dv">0</span>, min_bin <span class="op">=</span> <span class="dv">27</span>, max_bin <span class="op">=</span> <span class="dv">36</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>s3 <span class="op">=</span> call_from_fsa(fsa_files, s3_markers, liz120, val_plot<span class="op">=</span><span class="va">False</span>, intensity_plot<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>g3_a <span class="op">=</span> GenotypeResult()</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>g3_a.from_fsa_call(s3)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>g3_a.plotting_qc(<span class="st">'20180328-230857'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-14-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-14-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="example-of-adjusting-height" class="level1">
<h1>Example of adjusting height</h1>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># update height</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>fsa_files <span class="op">=</span> [<span class="st">'./raw_data/fsa/PA181114A/S1-20181106-730829_A10_SNaPshot50_POP7_E5_2018-11-23-16-40-43.fsa'</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>             <span class="st">'./raw_data/fsa/PA181114A/S2-20181106-730829_B10_SNaPshot50_POP7_E5_2018-11-23-16-40-43.fsa'</span>]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>liz120 <span class="op">=</span> [<span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">25</span>, <span class="dv">35</span>, <span class="dv">50</span>, <span class="dv">62</span>, <span class="dv">80</span>, <span class="dv">110</span>, <span class="dv">120</span>]</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>s4_markers <span class="op">=</span> generate_markers(peak_table)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>s4 <span class="op">=</span> call_from_fsa(fsa_files, s4_markers, liz120, val_plot<span class="op">=</span><span class="va">False</span>, intensity_plot<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-15-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-15-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>g4 <span class="op">=</span> GenotypeResult()</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>g4.from_fsa_call(s4)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>g4.plotting_qc(<span class="st">'20181106-730829'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-16-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-16-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>g4.allele_table(called_filter<span class="op">=</span><span class="va">False</span>).head(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample</th>
      <th>gene</th>
      <th>marker</th>
      <th>label</th>
      <th>panel</th>
      <th>direction</th>
      <th>base</th>
      <th>basetype</th>
      <th>min_bin</th>
      <th>max_bin</th>
      <th>min_height</th>
      <th>is_forward</th>
      <th>is_detected</th>
      <th>peak</th>
      <th>size</th>
      <th>height</th>
      <th>status</th>
      <th>message</th>
      <th>color</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>20181106-730829</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>G</td>
      <td>wildtype</td>
      <td>25</td>
      <td>35</td>
      <td>500</td>
      <td>1</td>
      <td>True</td>
      <td>25</td>
      <td>28.18</td>
      <td>891.0</td>
      <td>ok</td>
      <td></td>
      <td>blue</td>
    </tr>
    <tr>
      <th>1</th>
      <td>20181106-730829</td>
      <td>CYP2D6</td>
      <td>CYP2D6_001</td>
      <td>CYP2D6_14</td>
      <td>S1</td>
      <td>Forward</td>
      <td>A</td>
      <td>mutant</td>
      <td>27</td>
      <td>36</td>
      <td>500</td>
      <td>1</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>Peak(s) could not be detected. Please check pe...</td>
      <td>green</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20181106-730829</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>C</td>
      <td>wildtype</td>
      <td>28</td>
      <td>38</td>
      <td>500</td>
      <td>1</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>Peak(s) could not be detected. Please check pe...</td>
      <td>black</td>
    </tr>
    <tr>
      <th>3</th>
      <td>20181106-730829</td>
      <td>CYP2D6</td>
      <td>CYP2D6_002</td>
      <td>CYP2D6_10B</td>
      <td>S1</td>
      <td>Forward</td>
      <td>T</td>
      <td>mutant</td>
      <td>31</td>
      <td>37</td>
      <td>500</td>
      <td>1</td>
      <td>True</td>
      <td>25</td>
      <td>34.14</td>
      <td>542.0</td>
      <td>ok</td>
      <td></td>
      <td>red</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> marker_name <span class="kw">in</span> s4_markers:</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    marker <span class="op">=</span> s4_markers.get(marker_name)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> num, _ <span class="kw">in</span> <span class="bu">enumerate</span>(marker.alleles):</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>        marker.update_allele_config(allele_number<span class="op">=</span> num, min_height <span class="op">=</span><span class="dv">250</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>s4 <span class="op">=</span> call_from_fsa(fsa_files, s4_markers, liz120, val_plot<span class="op">=</span><span class="va">False</span>, intensity_plot<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>g4 <span class="op">=</span> GenotypeResult()</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>g4.from_fsa_call(s4)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>g4.plotting_qc(<span class="st">'20181106-730829'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-18-output-1.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-18-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-18-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="star_allele_calling_files/figure-html/cell-18-output-4.png" class="img-fluid"></p>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>